diff --git a/m4/getopt.m4 b/m4/getopt.m4
index af8bfd4c..0ccc413c 100644
--- a/m4/getopt.m4
+++ b/m4/getopt.m4
@@ -8,6 +8,9 @@ dnl with or without modifications, as long as this notice is preserved.
 
 # This version has been modified to reduce complexity since we only need
 # GNU getopt_long and do not care about replacing getopt.
+#
+# Pass gl_replace_getopt=yes (or any non-empty value instead of "yes") as
+# an argument to configure to force the use of the getopt_long replacement.
 
 # Check for a POSIX compliant getopt function with GNU extensions (such as
 # options with optional arguments) and the functions getopt_long,
@@ -23,8 +26,6 @@ AC_DEFUN([gl_FUNC_GETOPT_GNU],
 
 AC_DEFUN([gl_GETOPT_CHECK_HEADERS],
 [
-  gl_replace_getopt=
-
   if test -z "$gl_replace_getopt"; then
     AC_CHECK_HEADERS([getopt.h], [], [gl_replace_getopt=yes])
   fi
