diff --git a/src/scripts/Makefile.am b/src/scripts/Makefile.am
index 965f65ff..dbdca6fc 100644
--- a/src/scripts/Makefile.am
+++ b/src/scripts/Makefile.am
@@ -10,11 +10,16 @@ dist_man_MANS = xzdiff.1 xzgrep.1 xzmore.1
 
 install-exec-hook:
 	cd $(DESTDIR)$(bindir) && \
-	rm -f xzcmp xzegrep xzfgrep xzless lzcmp lzegrep lzfgrep lzless && \
+	rm -f xzcmp xzegrep xzfgrep xzless \
+		lzdiff lzgrep lzmore \
+		lzcmp lzegrep lzfgrep lzless && \
 	$(LN_S) xzdiff xzcmp && \
 	$(LN_S) xzgrep xzegrep && \
 	$(LN_S) xzgrep xzfgrep && \
 	$(LN_S) xzmore xzless && \
+	$(LN_S) xzdiff lzdiff && \
+	$(LN_S) xzgrep lzgrep && \
+	$(LN_S) xzmore lzmore && \
 	$(LN_S) xzdiff lzcmp && \
 	$(LN_S) xzgrep lzegrep && \
 	$(LN_S) xzgrep lzfgrep && \
@@ -23,11 +28,15 @@ install-exec-hook:
 install-data-hook:
 	cd $(DESTDIR)$(mandir)/man1 && \
 	rm -f xzcmp.1 xzegrep.1 xzfgrep.1 xzless.1 \
+		lzdiff.1 lzgrep.1 lzmore.1 \
 		lzcmp.1 lzegrep.1 lzfgrep.1 lzless.1 && \
 	$(LN_S) xzdiff.1 xzcmp.1 && \
 	$(LN_S) xzgrep.1 xzegrep.1 && \
 	$(LN_S) xzgrep.1 xzfgrep.1 && \
 	$(LN_S) xzmore.1 xzless.1 && \
+	$(LN_S) xzdiff.1 lzdiff.1 && \
+	$(LN_S) xzgrep.1 lzgrep.1 && \
+	$(LN_S) xzmore.1 lzmore.1 && \
 	$(LN_S) xzdiff.1 lzcmp.1 && \
 	$(LN_S) xzgrep.1 lzegrep.1 && \
 	$(LN_S) xzgrep.1 lzfgrep.1 && \
@@ -35,7 +44,10 @@ install-data-hook:
 
 uninstall-hook:
 	cd $(DESTDIR)$(bindir) && \
-	rm -f xzcmp xzegrep xzfgrep xzless lzcmp lzegrep lzfgrep lzless
+	rm -f xzcmp xzegrep xzfgrep xzless \
+		lzdiff lzgrep lzmore \
+		lzcmp lzegrep lzfgrep lzless
 	cd $(DESTDIR)$(mandir)/man1 && \
 	rm -f xzcmp.1 xzegrep.1 xzfgrep.1 xzless.1 \
+		lzdiff.1 lzgrep.1 lzmore.1 \
 		lzcmp.1 lzegrep.1 lzfgrep.1 lzless.1
