diff --git a/src/xz/file_io.c b/src/xz/file_io.c
index bf4462d9..08288ebb 100644
--- a/src/xz/file_io.c
+++ b/src/xz/file_io.c
@@ -738,6 +738,7 @@ io_open_dest_real(file_pair *pair)
 				message_error("%s: Refusing to write to "
 						"a DOS special file",
 						pair->dest_name);
+				free(pair->dest_name);
 				return true;
 			}
 
@@ -747,6 +748,7 @@ io_open_dest_real(file_pair *pair)
 				message_error("%s: Output file is the same "
 						"as the input file",
 						pair->dest_name);
+				free(pair->dest_name);
 				return true;
 			}
 		}
