diff --git a/src/liblzma/api/lzma/version.h b/src/liblzma/api/lzma/version.h
index 5064a781..36b7952d 100644
--- a/src/liblzma/api/lzma/version.h
+++ b/src/liblzma/api/lzma/version.h
@@ -22,7 +22,7 @@
  */
 #define LZMA_VERSION_MAJOR 4
 #define LZMA_VERSION_MINOR 999
-#define LZMA_VERSION_PATCH 8
+#define LZMA_VERSION_PATCH 9
 #define LZMA_VERSION_STABILITY LZMA_VERSION_STABILITY_BETA
 
 #ifndef LZMA_VERSION_COMMIT
