diff --git a/tests/test_compress.sh b/tests/test_compress.sh
index 05e3e8f7..9a3f25d8 100755
--- a/tests/test_compress.sh
+++ b/tests/test_compress.sh
@@ -102,15 +102,17 @@ do
 	test_xz -3
 	test_xz -4
 
+	# Disabled until Subblock format is stable.
+#		--subblock \
+#		--subblock=size=1 \
+#		--subblock=size=1,rle=1 \
+#		--subblock=size=1,rle=4 \
+#		--subblock=size=4,rle=4 \
+#		--subblock=size=8,rle=4 \
+#		--subblock=size=8,rle=8 \
+#		--subblock=size=4096,rle=12 \
+#	
 	for ARGS in \
-		--subblock \
-		--subblock=size=1 \
-		--subblock=size=1,rle=1 \
-		--subblock=size=1,rle=4 \
-		--subblock=size=4,rle=4 \
-		--subblock=size=8,rle=4 \
-		--subblock=size=8,rle=8 \
-		--subblock=size=4096,rle=12 \
 		--delta=dist=1 \
 		--delta=dist=4 \
 		--delta=dist=256 \
@@ -122,7 +124,9 @@ do
 		--sparc
 	do
 		test_xz $ARGS --lzma2=dict=64KiB,nice=32,mode=fast
-		test_xz --subblock $ARGS --lzma2=dict=64KiB,nice=32,mode=fast
+		
+		# Disabled until Subblock format is stable.
+		# test_xz --subblock $ARGS --lzma2=dict=64KiB,nice=32,mode=fast
 	done
 
 	echo
