diff --git a/CMakeLists.txt b/CMakeLists.txt
index 82cdde21..ac426c44 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1498,7 +1498,7 @@ if(BUILD_TESTING)
         #
         # Set the return code for skipped tests to match Automake convention.
         set_tests_properties("${TEST}" PROPERTIES
-            ENVIRONMENT "srcdir=${CMAKE_CURRENT_LIST_DIR}/tests"
+            ENVIRONMENT "srcdir=${CMAKE_CURRENT_SOURCE_DIR}/tests"
             SKIP_RETURN_CODE 77
         )
     endforeach()
