diff --git a/src/xz/file_io.c b/src/xz/file_io.c
index 3c3af0e1..4c52da56 100644
--- a/src/xz/file_io.c
+++ b/src/xz/file_io.c
@@ -73,8 +73,11 @@ io_init(void)
 static void
 io_unlink(const char *name, const struct stat *known_st)
 {
+#ifdef DOSLIKE
 	// On Windows, st_ino is meaningless, so don't bother testing it.
-#ifndef DOSLIKE
+	// Just silence a compiler warning.
+	(void)known_st;
+#else
 	struct stat new_st;
 
 	if (lstat(name, &new_st)
