diff --git a/src/common/sysdefs.h b/src/common/sysdefs.h
index 2fbcd593..97be4ee3 100644
--- a/src/common/sysdefs.h
+++ b/src/common/sysdefs.h
@@ -149,14 +149,8 @@ typedef unsigned char _Bool;
 #	define __bool_true_false_are_defined 1
 #endif
 
-// string.h should be enough but let's include strings.h too if it exists
-// since that shouldn't do any harm but may improve portability.
 #include <string.h>
 
-#ifdef HAVE_STRINGS_H
-#	include <strings.h>
-#endif
-
 // As of MSVC 2013, inline and restrict are supported with
 // non-standard keywords.
 #if defined(_WIN32) && defined(_MSC_VER)
