diff --git a/build-aux/ci_build.bash b/build-aux/ci_build.bash
index 7781a54b..7e9231fc 100755
--- a/build-aux/ci_build.bash
+++ b/build-aux/ci_build.bash
@@ -242,10 +242,6 @@ then
 
 		add_extra_option "$SMALL" "-DXZ_SMALL=ON" ""
 
-		if test -n "$CC" ; then
-			EXTRA_OPTIONS="$EXTRA_OPTIONS -DCMAKE_C_COMPILER=$CC"
-		fi
-
 		# Remove old cache file to clear previous settings.
 		rm -f "CMakeCache.txt"
 		cmake "$SRC_DIR/CMakeLists.txt" -B "$DEST_DIR" $EXTRA_OPTIONS -DXZ_CHECKS="$CHECK_TYPE" -G "Unix Makefiles"
