diff --git a/tests/files/README b/tests/files/README
index 84c9415c..e8f4160e 100644
--- a/tests/files/README
+++ b/tests/files/README
@@ -58,6 +58,9 @@
     good-single-none-empty_2.lzma is an empty file with implicit Copy
     filter and CRC32 as Check.
 
+    good-single-none-empty_3.lzma is an empty file with implicit Copy
+    filter, known Compressed Size, and no integrity Check.
+
     good-single-lzma-empty.lzma is an empty file with LZMA filter and no
     integrity Check.
 
@@ -85,6 +88,9 @@
     with 0x5D appended to the end of the file. 0x5D is the most common
     first byte of LZMA_Alone format file.
 
+    bad-single-none-empty.lzma is like good-single-none-empty_3.lzma but
+    with non-zero value in the Compressed Size field.
+
     bad-single-data_after_eopm_1.lzma has LZMA+Subblock, where the Subblock
     filter gives one byte of data to LZMA after LZMA has detected EOPM.
 
diff --git a/tests/files/bad-single-none-empty.lzma b/tests/files/bad-single-none-empty.lzma
new file mode 100644
index 00000000..3007e9a9
Binary files /dev/null and b/tests/files/bad-single-none-empty.lzma differ
diff --git a/tests/files/good-single-none-empty_3.lzma b/tests/files/good-single-none-empty_3.lzma
new file mode 100644
index 00000000..7e1d19b6
Binary files /dev/null and b/tests/files/good-single-none-empty_3.lzma differ
