diff --git a/THANKS b/THANKS
index ccbeff04..be1ee49c 100644
--- a/THANKS
+++ b/THANKS
@@ -80,6 +80,7 @@ has been important. :-) In alphabetical order:
   - Per Øyvind Karlsen
   - Firas Khalil Khana
   - Iouri Kharon
+  - Kim Jinyeong
   - Thomas Klausner
   - Richard Koch
   - Anton Kochkov
@@ -116,6 +117,7 @@ has been important. :-) In alphabetical order:
   - Étienne Mollier
   - Conley Moorhous
   - Dirk Müller
+  - Rainer Müller
   - Andrew Murray
   - Rafał Mużyło
   - Adrien Nader
