diff --git a/src/lzma/help.c b/src/lzma/help.c
index 3b9e3989..2e59f3b5 100644
--- a/src/lzma/help.c
+++ b/src/lzma/help.c
@@ -55,7 +55,7 @@ show_help(void)
 "  -k, --keep          keep (don't delete) input files\n"
 "  -f, --force         force overwrite of output file and (de)compress links\n"
 "  -c, --stdout        write to standard output and don't delete input files\n"
-"  -S, --suffix=.SUF   use suffix `.SUF' on compressed files instead of `.lzma'\n"
+"  -S, --suffix=.SUF   use the suffix `.SUF' on compressed files\n"
 "  -F, --format=FMT    file format to encode or decode; possible values are\n"
 "                      `auto' (default), `xz', `lzma', and `raw'\n"
 "      --files=[FILE]  read filenames to process from FILE; if FILE is\n"
@@ -81,11 +81,11 @@ show_help(void)
 "\n"
 "  --lzma1=[OPTS]      LZMA1 or LZMA2; OPTS is a comma-separated list of zero or\n"
 "  --lzma2=[OPTS]      more of the following options (valid values; default):\n"
-"                        dict=NUM   dictionary size in bytes (1 - 1GiB; 8MiB)\n"
+"                        dict=NUM   dictionary size (4KiB - 1536MiB; 8MiB)\n"
 "                        lc=NUM     number of literal context bits (0-4; 3)\n"
 "                        lp=NUM     number of literal position bits (0-4; 0)\n"
 "                        pb=NUM     number of position bits (0-4; 2)\n"
-"                        mode=MODE  compression mode (`fast' or `best'; `best')\n"
+"                        mode=MODE  compression mode (fast, normal; normal)\n"
 "                        nice=NUM   nice length of a match (2-273; 64)\n"
 "                        mf=NAME    match finder (hc3, hc4, bt2, bt3, bt4; bt4)\n"
 "                        depth=NUM  maximum search depth; 0=automatic (default)\n"
