diff --git a/src/liblzma/api/lzma/container.h b/src/liblzma/api/lzma/container.h
index 34cf6f48..86991add 100644
--- a/src/liblzma/api/lzma/container.h
+++ b/src/liblzma/api/lzma/container.h
@@ -96,6 +96,10 @@ typedef struct {
 	 * Set this to 0 to let liblzma choose the block size depending
 	 * on the compression options. For LZMA2 it will be 3*dict_size
 	 * or 1 MiB, whichever is more.
+	 *
+	 * For each thread, about 3 * block_size bytes of memory will be
+	 * allocated. This may change in later liblzma versions. If so,
+	 * the memory usage will probably be reduced, not increased.
 	 */
 	uint64_t block_size;
 
