diff --git a/NEWS b/NEWS
index 3ef11201..10634f00 100644
--- a/NEWS
+++ b/NEWS
@@ -2,6 +2,24 @@
 XZ Utils User-Visible Changes
 =============================
 
+5.0.2 (2011-04-01)
+
+    * LZMA2 decompressor now correctly accepts LZMA2 streams with no
+      uncompressed data. Previously it considered them corrupt. The
+      bug can affect applications that use raw LZMA2 streams. It is
+      very unlikely to affect .xz files because no compressor creates
+      .xz files with empty LZMA2 streams. (Empty .xz files are a
+      different thing than empty LZMA2 streams.)
+
+    * "xz --suffix=.foo filename.foo" now refuses to compress the
+      file due to it already having the suffix .foo. It was already
+      documented on the man page, but the code lacked the test.
+
+    * "xzgrep -l foo bar.xz" works now.
+
+    * Polish translation was added.
+
+
 5.0.1 (2011-01-29)
 
     * xz --force now (de)compresses files that have setuid, setgid,
