diff --git a/THANKS b/THANKS
index c5e34cd3..89c592c1 100644
--- a/THANKS
+++ b/THANKS
@@ -34,6 +34,7 @@ has been important. :-) In alphabetical order:
   - Vitaly Chikunov
   - Antoine Cœur
   - Felix Collin
+  - Ryan Colyer
   - Gabi Davar
   - İhsan Doğan
   - Chris Donawa
@@ -141,6 +142,7 @@ has been important. :-) In alphabetical order:
   - Bernhard Reutner-Fischer
   - Markus Rickert
   - Cristian Rodríguez
+  - Jeroen Roovers
   - Christian von Roques
   - Boud Roukema
   - Torsten Rupp
