diff --git a/src/xz/message.c b/src/xz/message.c
index 8fffa8c7..fcbf5967 100644
--- a/src/xz/message.c
+++ b/src/xz/message.c
@@ -1078,8 +1078,7 @@ message_help(bool long_help)
 	if (long_help)
 		puts(_(
 "  -M, --memory=NUM    use roughly NUM bytes of memory at maximum; 0 indicates\n"
-"                      the default setting, which depends on the operation mode\n"
-"                      and the amount of physical memory (RAM)"));
+"                      the default setting, which is 40 % of total RAM"));
 
 	if (long_help) {
 		puts(_(
