diff --git a/README b/README
index cce2666f..ca2867db 100644
--- a/README
+++ b/README
@@ -1,15 +1,25 @@
 
 XZ Utils
---------
-
-Warning
-
-    This is a beta version. Don't trust the files produced by
-    this version of the software - not even if the software can
-    uncompress the files properly! This is because the file format
-    isn't officially frozen yet although changes are very unlikely.
-
-    So please test a lot, but don't use for anything serious yet.
+========
+
+Important
+
+    This is a beta version. The .xz file format is now stable though,
+    which means that files created with the beta version will be
+    uncompressible with all future XZ Utils versions too (assuming
+    that there are no catastrophical bugs).
+
+    liblzma API is pretty stable now, although minor tweaks may still
+    be done if really needed. The ABI is not stable yet. The major
+    soname will be bumped right before the first stable release.
+    Probably it will be bumped to something like .so.5.0.0 because
+    some distributions using the alpha versions already had to use
+    other versions than .so.0.0.0.
+
+    Excluding the Doxygen style docs in libilzma API headers, the
+    documentation in this package (including the rest of this
+    README) is not very up to date, and may contain incorrect or
+    misleading information.
 
 
 Overview
