diff --git a/CMakeLists.txt b/CMakeLists.txt
index a6321672..ebf36b63 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1356,7 +1356,7 @@ endif()
 include(CMakePackageConfigHelpers)
 write_basic_package_version_file(
     "${CMAKE_CURRENT_BINARY_DIR}/liblzma-config-version.cmake"
-    VERSION "${liblzma_VERSION}"
+    VERSION "${xz_VERSION}"
     COMPATIBILITY SameMajorVersion)
 
 # Create liblzma-config.cmake. We use this spelling instead of
