diff --git a/CMakeLists.txt b/CMakeLists.txt
index e0af3e15..ecbb7b2d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -217,7 +217,7 @@ add_library(liblzma
     src/liblzma/api/lzma/vli.h
     src/liblzma/check/check.c
     src/liblzma/check/check.h
-    src/liblzma/check/crc_macros.h
+    src/liblzma/check/crc_common.h
     src/liblzma/common/block_util.c
     src/liblzma/common/common.c
     src/liblzma/common/common.h
diff --git a/src/liblzma/check/Makefile.inc b/src/liblzma/check/Makefile.inc
index dc011a3f..1b016356 100644
--- a/src/liblzma/check/Makefile.inc
+++ b/src/liblzma/check/Makefile.inc
@@ -12,7 +12,7 @@ EXTRA_DIST += \
 liblzma_la_SOURCES += \
 	check/check.c \
 	check/check.h \
-	check/crc_macros.h
+	check/crc_common.h
 
 if COND_CHECK_CRC32
 if COND_SMALL
diff --git a/src/liblzma/check/crc32_fast.c b/src/liblzma/check/crc32_fast.c
index eed73505..fa53e99b 100644
--- a/src/liblzma/check/crc32_fast.c
+++ b/src/liblzma/check/crc32_fast.c
@@ -17,7 +17,7 @@
 ///////////////////////////////////////////////////////////////////////////////
 
 #include "check.h"
-#include "crc_macros.h"
+#include "crc_common.h"
 
 
 // If you make any changes, do some benchmarking! Seemingly unrelated
diff --git a/src/liblzma/check/crc64_fast.c b/src/liblzma/check/crc64_fast.c
index 62c2c9bb..adca0584 100644
--- a/src/liblzma/check/crc64_fast.c
+++ b/src/liblzma/check/crc64_fast.c
@@ -77,7 +77,7 @@
 
 #ifdef CRC_GENERIC
 
-#include "crc_macros.h"
+#include "crc_common.h"
 
 
 #ifdef WORDS_BIGENDIAN
diff --git a/src/liblzma/check/crc_macros.h b/src/liblzma/check/crc_common.h
similarity index 96%
rename from src/liblzma/check/crc_macros.h
rename to src/liblzma/check/crc_common.h
index a7c21b76..f3ee205d 100644
--- a/src/liblzma/check/crc_macros.h
+++ b/src/liblzma/check/crc_common.h
@@ -1,6 +1,6 @@
 ///////////////////////////////////////////////////////////////////////////////
 //
-/// \file       crc_macros.h
+/// \file       crc_common.h
 /// \brief      Some endian-dependent macros for CRC32 and CRC64
 //
 //  Author:     Lasse Collin
