diff --git a/src/liblzma/api/lzma.h b/src/liblzma/api/lzma.h
index f4bfea1d..6a0c6bb8 100644
--- a/src/liblzma/api/lzma.h
+++ b/src/liblzma/api/lzma.h
@@ -12,13 +12,11 @@
  * includes a gzip-like command line tool named xz and some other tools.
  * XZ Utils is developed and maintained by Lasse Collin and Jia Tan.
  *
- * Major parts of liblzma are based on Igor Pavlov's public domain LZMA SDK
- * <https://7-zip.org/sdk.html>.
+ * Major parts of liblzma are based on code written by Igor Pavlov,
+ * specifically the LZMA SDK <https://7-zip.org/sdk.html>.
  *
- * The SHA-256 implementation is based on the public domain code found from
- * 7-Zip <https://7-zip.org/>, which has a modified version of the public
- * domain SHA-256 code found from Crypto++ <https://www.cryptopp.com/>.
- * The SHA-256 code in Crypto++ was written by Kevin Springle and Wei Dai.
+ * The SHA-256 implementation in liblzma is based on code written by
+ * Wei Dai in Crypto++ Library <https://www.cryptopp.com/>.
  *
  * liblzma is distributed under the BSD Zero Clause License (0BSD).
  */
