diff --git a/src/xz/file_io.c b/src/xz/file_io.c
index 978f19b1..d8a7597b 100644
--- a/src/xz/file_io.c
+++ b/src/xz/file_io.c
@@ -993,8 +993,8 @@ io_open_dest(file_pair *pair)
 /// \param      pair    File whose dest_fd should be closed
 /// \param      success If false, the file will be removed from the disk.
 ///
-/// \return     Zero if closing succeeds. On error, -1 is returned and
-///             error message printed.
+/// \return     If closing succeeds, false is returned. On error, an error
+///             message is printed and true is returned.
 static bool
 io_close_dest(file_pair *pair, bool success)
 {
diff --git a/src/xz/file_io.h b/src/xz/file_io.h
index ae7e2f38..f7b2f49e 100644
--- a/src/xz/file_io.h
+++ b/src/xz/file_io.h
@@ -177,6 +177,6 @@ extern bool io_pread(file_pair *pair, io_buf *buf, size_t size, uint64_t pos);
 /// \param      buf     Buffer containing the data to be written
 /// \param      size    Size of the buffer; must be at most IO_BUFFER_SIZE
 ///
-/// \return     On success, zero is returned. On error, -1 is returned
-///             and error message printed.
+/// \return     On success, false is returned. On error, error message
+///             is printed and true is returned.
 extern bool io_write(file_pair *pair, const io_buf *buf, size_t size);
