diff --git a/tests/files/README b/tests/files/README
index 841ec2c1..4a7d5f88 100644
--- a/tests/files/README
+++ b/tests/files/README
@@ -70,6 +70,14 @@
     good-single-delta-lzma.tiff.lzma is an image file that compresses
     better with Delta+LZMA than with plain LZMA.
 
+    good-single-x86-lzma.lzma uses the x86 filter (BCJ) and LZMA. The
+    uncompressed file is compress_prepared_bcj_x86 found from the tests
+    directory.
+
+    good-single-sparc-lzma.lzma uses the SPARC filter and LZMA. The
+    uncompressed file is compress_prepared_bcj_sparc found from the tests
+    directory.
+
     good-single-lzma-flush_1.lzma has a flush marker in the middle of
     the file, and no EOPM.
 
@@ -219,6 +227,9 @@
     decoder should abort before decoding any Data Blocks; all Multi-Block
     decoders must detect the file as corrupt at some point.
 
+    bad-multi-none-header_7.lzma has zero as the Size of Header Metadata
+    Block in the Header Metadata Block.
+
     bad-multi-none-block_1.lzma has wrong Uncompressed Size in the first
     Data Block. A sophisticated decoder should detect this error before
     producing any output, because it can see that the Uncompressed Size
diff --git a/tests/files/bad-multi-none-header_7.lzma b/tests/files/bad-multi-none-header_7.lzma
new file mode 100644
index 00000000..66b2d4b5
Binary files /dev/null and b/tests/files/bad-multi-none-header_7.lzma differ
diff --git a/tests/files/good-single-sparc-lzma.lzma b/tests/files/good-single-sparc-lzma.lzma
new file mode 100644
index 00000000..c464fbb3
Binary files /dev/null and b/tests/files/good-single-sparc-lzma.lzma differ
diff --git a/tests/files/good-single-x86-lzma.lzma b/tests/files/good-single-x86-lzma.lzma
new file mode 100644
index 00000000..04b9b2d4
Binary files /dev/null and b/tests/files/good-single-x86-lzma.lzma differ
