diff --git a/Doxyfile.in b/Doxyfile.in
index fa3c1e1d..386706ab 100644
--- a/Doxyfile.in
+++ b/Doxyfile.in
@@ -118,7 +118,7 @@ FULL_PATH_NAMES        = YES
 # If left blank the directory from which doxygen is run is used as the
 # path to strip.
 
-STRIP_FROM_PATH        =
+STRIP_FROM_PATH        = @top_srcdir@/src
 
 # The STRIP_FROM_INC_PATH tag can be used to strip a user-defined part of
 # the path mentioned in the documentation of a class, which tells
