diff --git a/README b/README
index 0676f6ad..39267967 100644
--- a/README
+++ b/README
@@ -9,7 +9,7 @@ XZ Utils
        1.3. Documentation for liblzma
     2. Version numbering
     3. Reporting bugs
-    4. Translating the xz tool
+    4. Translations
     5. Other implementations of the .xz format
     6. Contact information
 
@@ -192,9 +192,10 @@ XZ Utils
     system.
 
 
-4. Translating the xz tool
---------------------------
+4. Translations
+---------------
 
+    The xz command line tool and all man pages can be translated.
     The translations are handled via the Translation Project. If you
     wish to help translating xz, please join the Translation Project:
 
