diff --git a/po4a/po4a.conf b/po4a/po4a.conf
index c7a7d423..53425304 100644
--- a/po4a/po4a.conf
+++ b/po4a/po4a.conf
@@ -2,7 +2,7 @@
 # to get a new .po file. After translating the .po file, run
 # "update-po" again to generate the translated man pages.
 
-[po4a_langs] de fr ro uk ko pt_BR
+[po4a_langs] de fr ko pt_BR ro uk
 [po4a_paths] xz-man.pot $lang:$lang.po
 
 [type: man] ../src/xz/xz.1              $lang:man/$lang/xz.1
