diff --git a/src/lzmainfo/Makefile.am b/src/lzmainfo/Makefile.am
index 6662ecd5..5cb14a15 100644
--- a/src/lzmainfo/Makefile.am
+++ b/src/lzmainfo/Makefile.am
@@ -18,12 +18,14 @@ endif
 lzmainfo_CPPFLAGS = \
 	-DLOCALEDIR=\"$(localedir)\" \
 	-I$(top_srcdir)/src/common \
-	-I$(top_srcdir)/src/liblzma/api \
-	-I$(top_builddir)/lib
+	-I$(top_srcdir)/src/liblzma/api
 
 lzmainfo_LDADD = $(top_builddir)/src/liblzma/liblzma.la
 
 if COND_GNULIB
+lzmainfo_CPPFLAGS += \
+	-I$(top_builddir)/lib \
+	-I$(top_srcdir)/lib
 lzmainfo_LDADD += $(top_builddir)/lib/libgnu.a
 endif
 
diff --git a/src/xz/Makefile.am b/src/xz/Makefile.am
index 38d75ced..391d7145 100644
--- a/src/xz/Makefile.am
+++ b/src/xz/Makefile.am
@@ -50,12 +50,14 @@ endif
 xz_CPPFLAGS = \
 	-DLOCALEDIR=\"$(localedir)\" \
 	-I$(top_srcdir)/src/common \
-	-I$(top_srcdir)/src/liblzma/api \
-	-I$(top_builddir)/lib
+	-I$(top_srcdir)/src/liblzma/api
 
 xz_LDADD = $(top_builddir)/src/liblzma/liblzma.la
 
 if COND_GNULIB
+xz_CPPFLAGS += \
+	-I$(top_builddir)/lib \
+	-I$(top_srcdir)/lib
 xz_LDADD += $(top_builddir)/lib/libgnu.a
 endif
 
diff --git a/src/xzdec/Makefile.am b/src/xzdec/Makefile.am
index f300cc45..c2810479 100644
--- a/src/xzdec/Makefile.am
+++ b/src/xzdec/Makefile.am
@@ -21,11 +21,13 @@ endif
 xzdec_CPPFLAGS = \
 	-DTUKLIB_GETTEXT=0 \
 	-I$(top_srcdir)/src/common \
-	-I$(top_srcdir)/src/liblzma/api \
-	-I$(top_builddir)/lib
+	-I$(top_srcdir)/src/liblzma/api
 xzdec_LDADD = $(top_builddir)/src/liblzma/liblzma.la
 
 if COND_GNULIB
+xzdec_CPPFLAGS += \
+	-I$(top_builddir)/lib \
+	-I$(top_srcdir)/lib
 xzdec_LDADD += $(top_builddir)/lib/libgnu.a
 endif
 
