diff --git a/COPYING b/COPYING
index 8f5a85ba..43c90d05 100644
--- a/COPYING
+++ b/COPYING
@@ -39,11 +39,15 @@ XZ Utils Licensing
     the lack of copyright legally problematic, you have too many
     lawyers.
 
+    As usual, this software is provided "as is", without any warranty.
+
     If you copy significant amounts of public domain code from XZ Utils
     into your project, acknowledging this somewhere in your software is
-    polite, but naturally it is not legally required.
+    polite (especially if it is proprietary, non-free software), but
+    naturally it is not legally required. Here is an example of a good
+    notice to put into "about box" or into documentation:
 
-    As usual, this software is provided "as is", without any warranty.
+        This software includes code from XZ Utils <http://tukaani.org/xz/>.
 
     The following license texts are included in the following files:
       - COPYING.LGPLv2.1: GNU Lesser General Public License version 2.1
@@ -53,7 +57,8 @@ XZ Utils Licensing
     Note that the toolchain (compiler, linker etc.) may add some code
     pieces that are copyrighted. Thus, it is possible that e.g. liblzma
     binary wouldn't actually be in the public domain in its entirety
-    even though it contains no copyrighted code from this package.
+    even though it contains no copyrighted code from the XZ Utils source
+    package.
 
     If you have questions, don't hesitate to ask the author(s) for more
     information.
