diff --git a/configure.ac b/configure.ac
index b6f9f8b7..b8b52c9e 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1091,7 +1091,7 @@ uint32_t my_crc(uint32_t a, uint64_t b)
 # The C code is responsible for checking defined(__APPLE__) before using
 # sysctlbyname("hw.optional.armv8_crc32", ...).
 AS_IF([test "x$enable_arm64_crc32" = xyes], [
-	AC_CHECK_FUNCS([getauxval elf_aux_info sysctlbyname])
+	AC_CHECK_FUNCS([getauxval elf_aux_info sysctlbyname], [break])
 ])
 
 
