diff --git a/src/Makefile.am b/src/Makefile.am
index 9e08af5e..e1fcb5a9 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -27,8 +27,8 @@ EXTRA_DIST = \
 	common/tuklib_exit.h \
 	common/tuklib_gettext.h \
 	common/tuklib_integer.h \
-	common/tuklib_mbstr_fw.c \
 	common/tuklib_mbstr.h \
+	common/tuklib_mbstr_fw.c \
 	common/tuklib_mbstr_width.c \
 	common/tuklib_open_stdxxx.c \
 	common/tuklib_open_stdxxx.h \
