diff --git a/THANKS b/THANKS
index 426a2ed0..87140404 100644
--- a/THANKS
+++ b/THANKS
@@ -7,6 +7,7 @@ has been important. :-) In alphabetical order:
   - Mark Adler
   - H. Peter Anvin
   - Nelson H. F. Beebe
+  - Karl Berry
   - Anders F. Björklund
   - Emmanuel Blot
   - David Burklund
@@ -14,6 +15,7 @@ has been important. :-) In alphabetical order:
   - İsmail Dönmez
   - Mike Frysinger
   - Per Øyvind Karlsen
+  - Richard Koch
   - Ville Koskinen
   - Stephan Kulow
   - Peter Lawler
@@ -26,6 +28,7 @@ has been important. :-) In alphabetical order:
   - Mikko Pouru
   - Bernhard Reutner-Fischer
   - Christian von Roques
+  - Jukka Salmi
   - Alexandre Sauvé
   - Andreas Schwab
   - Dan Shechter
