diff --git a/Makefile.am b/Makefile.am
index e5b1265e..fa4b455d 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -79,7 +79,7 @@ manfiles = \
 dist-hook:
 	if test -d "$(srcdir)/.git" && type git > /dev/null 2>&1; then \
 		( cd "$(srcdir)" && git log --date=iso --stat \
-			b667a3ef6338a2c1db7b7706b1f6c99ea392221c^..HEAD ) \
+			6060f7dc76fd6c2a8a1f8e85d0e4d86bb78273e6^..HEAD ) \
 			> "$(distdir)/ChangeLog"; \
 	fi
 	if type groff > /dev/null 2>&1 && type ps2pdf > /dev/null 2>&1; then \
