diff --git a/Makefile.am b/Makefile.am
index 91c7c085..2ce74aa5 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -27,6 +27,11 @@ dist_doc_DATA = \
 	doc/xz-file-format.txt \
 	doc/lzma-file-format.txt
 
+examplesdir = $(docdir)/examples
+dist_examples_DATA = \
+	doc/examples/xz_pipe_comp.c \
+	doc/examples/xz_pipe_decomp.c
+
 EXTRA_DIST = \
 	extra \
 	dos \
diff --git a/windows/build.bash b/windows/build.bash
index 40808bc9..faec83b3 100644
--- a/windows/build.bash
+++ b/windows/build.bash
@@ -161,7 +161,7 @@ fi
 
 # Copy the headers, the .def file, and the docs.
 # They are the same for all architectures and builds.
-mkdir -pv pkg/{include/lzma,doc/manuals}
+mkdir -pv pkg/{include/lzma,doc/{manuals,examples}}
 txtcp pkg/include "" src/liblzma/api/lzma.h
 txtcp pkg/include/lzma "" src/liblzma/api/lzma/*.h
 txtcp pkg/doc "" src/liblzma/liblzma.def
@@ -169,6 +169,7 @@ txtcp pkg/doc .txt AUTHORS COPYING NEWS README THANKS TODO
 txtcp pkg/doc "" doc/*.txt windows/README-Windows.txt
 txtcp pkg/doc/manuals "" doc/man/txt/{xz,xzdec,lzmainfo}.txt
 cp -v doc/man/pdf-*/{xz,xzdec,lzmainfo}-*.pdf pkg/doc/manuals
+txtcp pkg/doc/examples "" doc/examples/*
 
 if [ -f windows/COPYING-Windows.txt ]; then
 	txtcp pkg/doc "" windows/COPYING-Windows.txt
