diff --git a/src/xz/args.c b/src/xz/args.c
index 393ba6aa..f37f8024 100644
--- a/src/xz/args.c
+++ b/src/xz/args.c
@@ -357,7 +357,7 @@ parse_real(args_info *args, int argc, char **argv)
 		case OPT_FILES0:
 			if (args->files_name != NULL)
 				message_fatal(_("Only one file can be "
-						"specified with `--files'"
+						"specified with `--files' "
 						"or `--files0'."));
 
 			if (optarg == NULL) {
