diff --git a/src/liblzma/api/lzma.h b/src/liblzma/api/lzma.h
index 4d8e3994..f38513d3 100644
--- a/src/liblzma/api/lzma.h
+++ b/src/liblzma/api/lzma.h
@@ -10,7 +10,7 @@
  *
  * liblzma is part of XZ Utils <https://tukaani.org/xz/>. XZ Utils includes
  * a gzip-like command line tool named xz and some other tools. XZ Utils
- * is developed and maintained by Lasse Collin.
+ * is developed and maintained by Lasse Collin and Jia Tan.
  *
  * Major parts of liblzma are based on Igor Pavlov's public domain LZMA SDK
  * <https://7-zip.org/sdk.html>.
