diff --git a/THANKS b/THANKS
index c8eb1186..ccbeff04 100644
--- a/THANKS
+++ b/THANKS
@@ -36,6 +36,7 @@ has been important. :-) In alphabetical order:
   - Felix Collin
   - Ryan Colyer
   - Marcus Comstedt
+  - Vincent Cruz
   - Gabi Davar
   - İhsan Doğan
   - Chris Donawa
@@ -114,6 +115,7 @@ has been important. :-) In alphabetical order:
   - Nathan Moinvaziri
   - Étienne Mollier
   - Conley Moorhous
+  - Dirk Müller
   - Andrew Murray
   - Rafał Mużyło
   - Adrien Nader
@@ -123,6 +125,7 @@ has been important. :-) In alphabetical order:
   - Jonathan Nieder
   - Asgeir Storesund Nilsen
   - Andre Noll
+  - Ruarí Ødegaard
   - Peter O'Gorman
   - Dimitri Papadopoulos Orfanos
   - Daniel Packard
@@ -139,6 +142,7 @@ has been important. :-) In alphabetical order:
   - Rich Prohaska
   - Trần Ngọc Quân
   - Pavel Raiskup
+  - Matthieu Rakotojaona
   - Ole André Vadla Ravnås
   - Eric S. Raymond
   - Robert Readman
