diff --git a/src/xz/coder.c b/src/xz/coder.c
index a4dbf3a0..c7898688 100644
--- a/src/xz/coder.c
+++ b/src/xz/coder.c
@@ -237,9 +237,9 @@ coder_set_compression_settings(void)
 	if (memory_usage <= memory_limit)
 		return;
 
-	// If --no-auto-adjust was used or we didn't find LZMA1 or
+	// If --no-adjust was used or we didn't find LZMA1 or
 	// LZMA2 as the last filter, give an error immediately.
-	// --format=raw implies --no-auto-adjust.
+	// --format=raw implies --no-adjust.
 	if (!opt_auto_adjust || opt_format == FORMAT_RAW)
 		memlimit_too_small(memory_usage);
 
