diff --git a/src/common/mythread.h b/src/common/mythread.h
index 4a885474..7a68da8f 100644
--- a/src/common/mythread.h
+++ b/src/common/mythread.h
@@ -25,7 +25,7 @@
 #ifdef MYTHREAD_ENABLED
 
 ////////////////////////////////////////
-// Shared betewen all threading types //
+// Shared between all threading types //
 ////////////////////////////////////////
 
 // Locks a mutex for a duration of a block.
diff --git a/src/liblzma/check/crc32_fast.c b/src/liblzma/check/crc32_fast.c
index 94da8559..3de02638 100644
--- a/src/liblzma/check/crc32_fast.c
+++ b/src/liblzma/check/crc32_fast.c
@@ -20,7 +20,7 @@
 #include "crc_macros.h"
 
 
-// If you make any changes, do some bench marking! Seemingly unrelated
+// If you make any changes, do some benchmarking! Seemingly unrelated
 // changes can very easily ruin the performance (and very probably is
 // very compiler dependent).
 extern LZMA_API(uint32_t)
