diff --git a/src/scripts/xzgrep.in b/src/scripts/xzgrep.in
index 7ab617b9..00242377 100644
--- a/src/scripts/xzgrep.in
+++ b/src/scripts/xzgrep.in
@@ -126,6 +126,10 @@ while test $# -ne 0; do
   grep="$grep $option$optarg"
 done
 
+if test $files_with_matches -eq 1 || test $files_without_matches -eq 1; then
+  grep="$grep -q"
+fi
+
 eval "set -- $operands "'${1+"$@"}'
 
 if test $have_pat -eq 0; then
@@ -157,9 +161,9 @@ for i; do
     exec 5>&1
     ($uncompress -- "$i" 5>&-; echo $? >&5) 3>&- |
     if test $files_with_matches -eq 1; then
-      eval "$grep" >/dev/null && { printf '%s\n' "$i" || exit 2; }
+      eval "$grep" && { printf '%s\n' "$i" || exit 2; }
     elif test $files_without_matches -eq 1; then
-      eval "$grep" >/dev/null || {
+      eval "$grep" || {
         r=$?
         if test $r -eq 1; then
           printf '%s\n' "$i" || r=2
