diff --git a/tests/ossfuzz/fuzz_decode_stream.c b/tests/ossfuzz/fuzz_decode_stream.c
index 158cf4d6..d7860611 100644
--- a/tests/ossfuzz/fuzz_decode_stream.c
+++ b/tests/ossfuzz/fuzz_decode_stream.c
@@ -21,7 +21,7 @@ extern int
 LLVMFuzzerTestOneInput(const uint8_t *inbuf, size_t inbuf_size)
 {
 	lzma_stream strm = LZMA_STREAM_INIT;
-	// Initialize a LZMA decoder using the memory usage limit
+	// Initialize a .xz decoder using the memory usage limit
 	// defined in fuzz_common.h
 	//
 	// Enable support for concatenated .xz files which is used when
