diff --git a/CMakeLists.txt b/CMakeLists.txt
index 032f4b37..a48a5fcd 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2051,9 +2051,9 @@ endif()
 #############################################################################
 
 if(UNIX)
-    option(ENABLE_DOXYGEN "Use Doxygen to generate liblzma API docs" OFF)
+    option(XZ_DOXYGEN "Use Doxygen to generate liblzma API docs" OFF)
 
-    if (ENABLE_DOXYGEN)
+    if (XZ_DOXYGEN)
         file(MAKE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/doc")
 
         add_custom_command(
