diff --git a/src/liblzma/common/info.c b/src/liblzma/common/info.c
index aee25463..b05bd031 100644
--- a/src/liblzma/common/info.c
+++ b/src/liblzma/common/info.c
@@ -385,7 +385,10 @@ lzma_info_metadata_set(lzma_info *info, lzma_allocator *allocator,
 
 		const lzma_ret ret = lzma_info_index_set(
 				info, allocator, metadata->index, eat_index);
-		metadata->index = NULL;
+
+		if (eat_index)
+			metadata->index = NULL;
+
 		if (ret != LZMA_OK)
 			return ret;
 
