diff --git a/doxygen/Doxyfile b/doxygen/Doxyfile
index 632a0b89..2bb2ac51 100644
--- a/doxygen/Doxyfile
+++ b/doxygen/Doxyfile
@@ -39,3 +39,7 @@ PREDEFINED             = LZMA_API(type)=type \
                          tuklib_attr_noreturn= \
                          lzma_attribute(attr)= \
                          lzma_attr_alloc_size(size)=
+
+# Debian and Ubuntu patch Doxygen so that HAVE_DOT = YES is the default.
+# Set HAVE_DOT explicitly to get consistent behavior across distributions.
+HAVE_DOT               = NO
