diff --git a/THANKS b/THANKS
index 8cef6d22..b638ffd7 100644
--- a/THANKS
+++ b/THANKS
@@ -20,6 +20,7 @@ has been important. :-) In alphabetical order:
   - Jakub Bogusz
   - Adam Borowski
   - Maarten Bosmans
+  - Roel Bouckaert
   - Lukas Braune
   - Benjamin Buch
   - Trent W. Buck
@@ -145,6 +146,7 @@ has been important. :-) In alphabetical order:
   - Igor Pavlov
   - Diego Elio Pettenò
   - Elbert Pol
+  - Guiorgy Potskhishvili
   - Mikko Pouru
   - Frank Prochnow
   - Rich Prohaska
@@ -206,6 +208,7 @@ has been important. :-) In alphabetical order:
   - Ralf Wildenhues
   - Charles Wilson
   - Lars Wirzenius
+  - Vincent Wixsom
   - Pilorz Wojciech
   - Chien Wong
   - Xi Ruoyao
