diff --git a/INSTALL b/INSTALL
index 37c40055..ba187bbb 100644
--- a/INSTALL
+++ b/INSTALL
@@ -90,21 +90,15 @@ XZ Utils Installation
 
 1.2.3. MINIX 3
 
-    The default install of MINIX 3 includes Amsterdam Compiler Kit (ACK),
-    which doesn't support C99. Install GCC to compile XZ Utils.
+    Version 3.3.0 and later are supported.
 
-    MINIX 3.1.8 and older have bugs in /usr/include/stdint.h, which has
-    to be patched before XZ Utils can be compiled correctly. See
-    <http://gforge.cs.vu.nl/gf/project/minix/tracker/?action=TrackerItemEdit&tracker_item_id=537>.
+    Multithreading isn't supported because MINIX 3 doesn't have
+    pthreads. The option --disable-threads must be passed to configure
+    as this isn't autodetected.
 
-    MINIX 3.2.0 and later use a different libc and aren't affected by
-    the above bug.
-
-    XZ Utils doesn't have code to detect the amount of physical RAM and
-    number of CPU cores on MINIX 3.
-
-    See section 5.4 in this file about symbol visibility warnings (you
-    may want to pass gl_cv_cc_visibility=no to configure).
+    Note that disabling threads causes "make check" to show a few tests
+    as skipped ("SKIP"). It's only due to a few threading-dependent
+    subtests are skipped. See the matching tests/test_*.log files.
 
 
 1.2.4. OpenVMS
