diff --git a/lib/Makefile.am b/lib/Makefile.am
index 46e3a4a9..0ce2a676 100644
--- a/lib/Makefile.am
+++ b/lib/Makefile.am
@@ -15,19 +15,11 @@
 
 ## Not using gnulib-tool, at least for now. Less mess this way.
 
-## We need two builds of libgnu: one with NLS and one without.
-## This is because lzma uses NLS but lzmadec doesn't, while
-## both need GNU getopt_long().
-noinst_LIBRARIES = libgnu.a libgnu_nls.a
+noinst_LIBRARIES = libgnu.a
 
 libgnu_a_SOURCES =
 libgnu_a_DEPENDENCIES = $(LIBOBJS)
 libgnu_a_LIBADD = $(LIBOBJS)
-libgnu_a_CPPFLAGS = -DDISABLE_NLS=1
-
-libgnu_nls_a_SOURCES =
-libgnu_nls_a_DEPENDENCIES = $(LIBOBJS)
-libgnu_nls_a_LIBADD = $(LIBOBJS)
 
 EXTRA_DIST = gettext.h getopt_.h getopt.c getopt1.c getopt_int.h
 BUILT_SOURCES = $(GETOPT_H)
diff --git a/src/lzma/Makefile.am b/src/lzma/Makefile.am
index 5fbd3358..84e4d34d 100644
--- a/src/lzma/Makefile.am
+++ b/src/lzma/Makefile.am
@@ -59,5 +59,5 @@ lzma_LDADD = \
 	@PTHREAD_LIBS@
 
 if COND_GNULIB
-lzma_LDADD += @top_builddir@/lib/libgnu_nls.a
+lzma_LDADD += @top_builddir@/lib/libgnu.a
 endif
diff --git a/src/lzmadec/Makefile.am b/src/lzmadec/Makefile.am
index 63e391d5..650eb1c4 100644
--- a/src/lzmadec/Makefile.am
+++ b/src/lzmadec/Makefile.am
@@ -20,7 +20,9 @@ lzmadec_CPPFLAGS = \
 	-I@top_srcdir@/src/liblzma/api \
 	-I@top_builddir@/lib
 lzmadec_LDFLAGS = -static
-lzmadec_LDADD = @top_builddir@/src/liblzma/liblzma.la
+lzmadec_LDADD = \
+	@top_builddir@/src/liblzma/liblzma.la \
+	@LTLIBINTL@
 
 if COND_GNULIB
 lzmadec_LDADD += @top_builddir@/lib/libgnu.a
