diff --git a/src/liblzma/rangecoder/price_table_gen.c b/src/liblzma/rangecoder/price_table_gen.c
index 04703e4f..946d8215 100644
--- a/src/liblzma/rangecoder/price_table_gen.c
+++ b/src/liblzma/rangecoder/price_table_gen.c
@@ -43,7 +43,7 @@ main(void)
 		if (i % 8 == 0)
 			printf("\n\t");
 
-		printf("% 4" PRIu32, lzma_rc_prob_prices[i]);
+		printf("%4" PRIu32, lzma_rc_prob_prices[i]);
 
 		if (i != array_size - 1)
 			printf(",");
