diff --git a/src/liblzma/common/easy_preset.h b/src/liblzma/common/easy_preset.h
index d0ca1a62..4ef6d044 100644
--- a/src/liblzma/common/easy_preset.h
+++ b/src/liblzma/common/easy_preset.h
@@ -9,6 +9,9 @@
 //
 ///////////////////////////////////////////////////////////////////////////////
 
+#ifndef LZMA_EASY_PRESET_H
+#define LZMA_EASY_PRESET_H
+
 #include "common.h"
 
 
@@ -29,3 +32,5 @@ typedef struct {
 /// Set *easy to the settings given by the preset. Returns true on error,
 /// false on success.
 extern bool lzma_easy_preset(lzma_options_easy *easy, uint32_t preset);
+
+#endif
