diff --git a/THANKS b/THANKS
index 5f8ca492..bbdc079e 100644
--- a/THANKS
+++ b/THANKS
@@ -42,9 +42,11 @@ has been important. :-) In alphabetical order:
   - Michael Fox
   - Mike Frysinger
   - Daniel Richard G.
+  - Tomasz Gajc
   - Bjarni Ingi Gislason
   - John Paul Adrian Glaubitz
   - Bill Glessner
+  - Michał Górny
   - Jason Gorski
   - Juan Manuel Guerrero
   - Diederik de Haas
@@ -53,6 +55,7 @@ has been important. :-) In alphabetical order:
   - Vincenzo Innocente
   - Peter Ivanov
   - Nicholas Jackson
+  - Sam James
   - Jouk Jansen
   - Jun I Jin
   - Kiyoshi Kanazawa
