diff --git a/THANKS b/THANKS
index 4301f205..38ce0c1d 100644
--- a/THANKS
+++ b/THANKS
@@ -36,6 +36,7 @@ has been important. :-) In alphabetical order:
   - Michael Fox
   - Mike Frysinger
   - Daniel Richard G.
+  - Bjarni Ingi Gislason
   - Bill Glessner
   - Jason Gorski
   - Juan Manuel Guerrero
@@ -88,9 +89,10 @@ has been important. :-) In alphabetical order:
   - Trần Ngọc Quân
   - Pavel Raiskup
   - Ole André Vadla Ravnås
+  - Eric S. Raymond
   - Robert Readman
   - Bernhard Reutner-Fischer
-  - Eric S. Raymond
+  - Markus Rickert
   - Cristian Rodríguez
   - Christian von Roques
   - Torsten Rupp
