diff --git a/cmake/tuklib_cpucores.cmake b/cmake/tuklib_cpucores.cmake
index e5e9c34d..05f3ceef 100644
--- a/cmake/tuklib_cpucores.cmake
+++ b/cmake/tuklib_cpucores.cmake
@@ -9,6 +9,7 @@
 #############################################################################
 
 include("${CMAKE_CURRENT_LIST_DIR}/tuklib_common.cmake")
+include(CMakePushCheckState)
 include(CheckCSourceCompiles)
 include(CheckIncludeFile)
 
@@ -76,6 +77,7 @@ function(tuklib_cpucores_internal_check)
     #
     # We test sysctl() first and intentionally break the sysctl() test on QNX
     # so that sysctl() is never used on QNX.
+    cmake_push_check_state()
     check_include_file(sys/param.h HAVE_SYS_PARAM_H)
     if(HAVE_SYS_PARAM_H)
         list(APPEND CMAKE_REQUIRED_DEFINITIONS -DHAVE_SYS_PARAM_H)
@@ -103,6 +105,7 @@ function(tuklib_cpucores_internal_check)
             }
         "
         TUKLIB_CPUCORES_SYSCTL)
+    cmake_pop_check_state()
     if(TUKLIB_CPUCORES_SYSCTL)
         if(HAVE_SYS_PARAM_H)
             set(TUKLIB_CPUCORES_DEFINITIONS
diff --git a/cmake/tuklib_physmem.cmake b/cmake/tuklib_physmem.cmake
index e4888ebe..d4d3f3d2 100644
--- a/cmake/tuklib_physmem.cmake
+++ b/cmake/tuklib_physmem.cmake
@@ -12,6 +12,7 @@
 #############################################################################
 
 include("${CMAKE_CURRENT_LIST_DIR}/tuklib_common.cmake")
+include(CMakePushCheckState)
 include(CheckCSourceCompiles)
 include(CheckIncludeFile)
 
@@ -76,11 +77,11 @@ function(tuklib_physmem_internal_check)
     endif()
 
     # sysctl()
+    cmake_push_check_state()
     check_include_file(sys/param.h HAVE_SYS_PARAM_H)
     if(HAVE_SYS_PARAM_H)
         list(APPEND CMAKE_REQUIRED_DEFINITIONS -DHAVE_SYS_PARAM_H)
     endif()
-
     check_c_source_compiles("
             #ifdef HAVE_SYS_PARAM_H
             #   include <sys/param.h>
@@ -96,6 +97,7 @@ function(tuklib_physmem_internal_check)
             }
         "
         TUKLIB_PHYSMEM_SYSCTL)
+    cmake_pop_check_state()
     if(TUKLIB_PHYSMEM_SYSCTL)
         if(HAVE_SYS_PARAM_H)
             set(TUKLIB_PHYSMEM_DEFINITIONS
