diff --git a/src/liblzma/check/crc_common.h b/src/liblzma/check/crc_common.h
index 0dd4faec..51ddd9d5 100644
--- a/src/liblzma/check/crc_common.h
+++ b/src/liblzma/check/crc_common.h
@@ -89,7 +89,9 @@
 	// 16 bytes and CLMUL for bigger inputs. It saves a little in code
 	// size since the special cases for 0-16-byte inputs will be omitted
 	// from the CLMUL code.
-#	define CRC_USE_GENERIC_FOR_SMALL_INPUTS 1
+#	ifndef CRC_USE_IFUNC
+#		define CRC_USE_GENERIC_FOR_SMALL_INPUTS 1
+#	endif
 */
 
 #	if defined(_MSC_VER)
