diff --git a/.gitignore b/.gitignore
index efb7b009..ef4754db 100644
--- a/.gitignore
+++ b/.gitignore
@@ -62,10 +62,11 @@ coverage
 /tests/test_block_header
 /tests/test_check
 /tests/test_filter_flags
+/tests/test_filter_str
 /tests/test_hardware
 /tests/test_index
 /tests/test_index_hash
-/test/test_lzip_decoder
+/tests/test_lzip_decoder
 /tests/test_memlimit
 /tests/test_stream_flags
 /tests/test_vli
