diff --git a/THANKS b/THANKS
index 0b438f4d..01e6d0bf 100644
--- a/THANKS
+++ b/THANKS
@@ -24,6 +24,7 @@ has been important. :-) In alphabetical order:
   - Milo Casagrande
   - Marek Černocký
   - Tomer Chachamu
+  - Vitaly Chikunov
   - Antoine Cœur
   - Gabi Davar
   - Chris Donawa
@@ -71,6 +72,7 @@ has been important. :-) In alphabetical order:
   - Gregory Margo
   - Julien Marrec
   - Martin Matuška
+  - Ivan A. Melnikov
   - Jim Meyering
   - Arkadiusz Miskiewicz
   - Nathan Moinvaziri
