diff --git a/README b/README
index cdba8870..32d5084c 100644
--- a/README
+++ b/README
@@ -95,9 +95,9 @@ XZ Utils
 
     The command-line tools are documented as man pages. In source code
     releases (and possibly also in some binary packages), the man pages
-    are also provided in plain text (ASCII only) and PDF formats in the
-    directory "doc/man" to make the man pages more accessible to those
-    whose operating system doesn't provide an easy way to view man pages.
+    are also provided in plain text (ASCII only) format in the directory
+    "doc/man" to make the man pages more accessible to those whose
+    operating system doesn't provide an easy way to view man pages.
 
 
 1.3. Documentation for liblzma
