diff --git a/windows/vs2019/config.h b/windows/vs2019/config.h
index 25ac18e7..ef921e80 100644
--- a/windows/vs2019/config.h
+++ b/windows/vs2019/config.h
@@ -1,4 +1,4 @@
-/* config.h for compiling liblzma (*not* the whole XZ Utils) with MSVC 2017 */
+/* config.h for compiling liblzma (*not* the whole XZ Utils) with MSVC 2019 */
 
 /* Prefix for symbols exported by tuklib_*.c files */
 #define TUKLIB_SYMBOL_PREFIX lzma_
