diff --git a/THANKS b/THANKS
index bcf3d4ff..b72779b9 100644
--- a/THANKS
+++ b/THANKS
@@ -152,9 +152,9 @@ has been important. :-) In alphabetical order:
   - Dan Stromberg
   - Jia Tan
   - Vincent Torri
+  - Alexey Tourbin
   - Paul Townsend
   - Mohammed Adnène Trojette
-  - Alexey Tourbin
   - Taiki Tsunekawa
   - Maksym Vatsyk
   - Loganaden Velvindron
@@ -170,8 +170,8 @@ has been important. :-) In alphabetical order:
   - Ralf Wildenhues
   - Charles Wilson
   - Lars Wirzenius
-  - Chien Wong
   - Pilorz Wojciech
+  - Chien Wong
   - Ryan Young
   - Andreas Zieringer
 
