diff --git a/THANKS b/THANKS
index aee1183a..7ce17501 100644
--- a/THANKS
+++ b/THANKS
@@ -113,6 +113,7 @@ has been important. :-) In alphabetical order:
   - Rafał Mużyło
   - Adrien Nader
   - Evan Nemerson
+  - Alexander Neumann
   - Hongbo Ni
   - Jonathan Nieder
   - Andre Noll
@@ -121,6 +122,7 @@ has been important. :-) In alphabetical order:
   - Daniel Packard
   - Filip Palian
   - Peter Pallinger
+  - Kai Pastor
   - Rui Paulo
   - Igor Pavlov
   - Diego Elio Pettenò
@@ -145,6 +147,7 @@ has been important. :-) In alphabetical order:
   - Alexandre Sauvé
   - Benno Schulenberg
   - Andreas Schwab
+  - Eli Schwartz
   - Bhargava Shastry
   - Dan Shechter
   - Stuart Shelton
