diff --git a/Makefile.am b/Makefile.am
index 678d9eb3..79cf1954 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -14,7 +14,7 @@ endif
 
 SUBDIRS += src po tests
 
-doc_DATA = \
+dist_doc_DATA = \
 	AUTHORS \
 	COPYING \
 	COPYING.GPLv2 \
@@ -28,7 +28,6 @@ doc_DATA = \
 	doc/lzma-file-format.txt
 
 EXTRA_DIST = \
-	doc \
 	extra \
 	dos \
 	windows \
@@ -55,13 +54,13 @@ manfiles = \
 # Create ChangeLog from output of "git log --date=iso --stat".
 # Convert the man pages to PDF and plain text (ASCII only) formats.
 dist-hook:
-	test -d "$(srcdir)/.git" && type git > /dev/null 2>&1 \
-		&& ( cd "$(srcdir)" && git log --date=iso --stat ) \
-			> "$(distdir)/ChangeLog"
+	if test -d "$(srcdir)/.git" && type git > /dev/null 2>&1; then \
+		( cd "$(srcdir)" && git log --date=iso --stat ) \
+			> "$(distdir)/ChangeLog"; \
+	fi
 	if type groff > /dev/null 2>&1 && type ps2pdf > /dev/null 2>&1; then \
 		dest="$(distdir)/doc/man" && \
-		mkdir "$$dest" && \
-		mkdir "$$dest/pdf-a4" "$$dest/pdf-letter" "$$dest/txt" && \
+		$(MKDIR_P) "$$dest/pdf-a4" "$$dest/pdf-letter" "$$dest/txt" && \
 		for FILE in $(manfiles); do \
 			BASE=`basename $$FILE .1` && \
 			groff -man -t -Tps -P-pa4 < "$(srcdir)/$$FILE" \
