diff --git a/src/xz/message.c b/src/xz/message.c
index c73099e6..38cce4a1 100644
--- a/src/xz/message.c
+++ b/src/xz/message.c
@@ -833,10 +833,13 @@ message_strm(lzma_ret code)
 	case LZMA_STREAM_END:
 	case LZMA_GET_CHECK:
 	case LZMA_PROG_ERROR:
-		return _("Internal error (bug)");
+		// Without "default", compiler will warn if new constants
+		// are added to lzma_ret, it is not too easy to forget to
+		// add the new constants to this function.
+		break;
 	}
 
-	return NULL;
+	return _("Internal error (bug)");
 }
 
 
