diff --git a/src/liblzma/check/crc32_x86.S b/src/liblzma/check/crc32_x86.S
index 4737a544..ff4d8a38 100644
--- a/src/liblzma/check/crc32_x86.S
+++ b/src/liblzma/check/crc32_x86.S
@@ -67,7 +67,7 @@ init_table(void)
  * Solaris assembler doesn't have .p2align, and Darwin uses .align
  * differently than GNU/Linux and Solaris.
  */
-#if defined(__MACH__) || defined(__MSDOS__)
+#if defined(__APPLE__) || defined(__MSDOS__)
 #	define ALIGN(pow2, abs) .align pow2
 #else
 #	define ALIGN(pow2, abs) .align abs
@@ -76,7 +76,7 @@ init_table(void)
 	.text
 	.globl	LZMA_CRC32
 
-#if !defined(__MACH__) && !defined(_WIN32) && !defined(__CYGWIN__) \
+#if !defined(__APPLE__) && !defined(_WIN32) && !defined(__CYGWIN__) \
 		&& !defined(__MSDOS__)
 	.type	LZMA_CRC32, @function
 #endif
@@ -119,7 +119,7 @@ LZMA_CRC32:
 		|| (defined(_WIN32) || defined(__CYGWIN__))
 	/* Not PIC */
 	movl	$ LZMA_CRC32_TABLE, %ebx
-#elif defined(__MACH__)
+#elif defined(__APPLE__)
 	/* Mach-O */
 	call	.L_get_pc
 .L_pic:
@@ -275,7 +275,7 @@ LZMA_CRC32:
 	ret
 #endif
 
-#if defined(__MACH__) && (defined(PIC) || defined(__PIC__))
+#if defined(__APPLE__) && (defined(PIC) || defined(__PIC__))
 	/* Mach-O PIC */
 	.section __IMPORT,__pointers,non_lazy_symbol_pointers
 .L_lzma_crc32_table$non_lazy_ptr:
diff --git a/src/liblzma/check/crc64_x86.S b/src/liblzma/check/crc64_x86.S
index 9a63726e..e095ce1a 100644
--- a/src/liblzma/check/crc64_x86.S
+++ b/src/liblzma/check/crc64_x86.S
@@ -57,7 +57,7 @@ init_table(void)
  * Solaris assembler doesn't have .p2align, and Darwin uses .align
  * differently than GNU/Linux and Solaris.
  */
-#if defined(__MACH__) || defined(__MSDOS__)
+#if defined(__APPLE__) || defined(__MSDOS__)
 #	define ALIGN(pow2, abs) .align pow2
 #else
 #	define ALIGN(pow2, abs) .align abs
@@ -66,7 +66,7 @@ init_table(void)
 	.text
 	.globl	LZMA_CRC64
 
-#if !defined(__MACH__) && !defined(_WIN32) && !defined(__CYGWIN__) \
+#if !defined(__APPLE__) && !defined(_WIN32) && !defined(__CYGWIN__) \
 		&& !defined(__MSDOS__)
 	.type	LZMA_CRC64, @function
 #endif
@@ -110,7 +110,7 @@ LZMA_CRC64:
 		|| (defined(_WIN32) || defined(__CYGWIN__))
 	/* Not PIC */
 	movl	$ LZMA_CRC64_TABLE, %ebx
-#elif defined(__MACH__)
+#elif defined(__APPLE__)
 	/* Mach-O */
 	call	.L_get_pc
 .L_pic:
@@ -258,7 +258,7 @@ LZMA_CRC64:
 	ret
 #endif
 
-#if defined(__MACH__) && (defined(PIC) || defined(__PIC__))
+#if defined(__APPLE__) && (defined(PIC) || defined(__PIC__))
 	/* Mach-O PIC */
 	.section __IMPORT,__pointers,non_lazy_symbol_pointers
 .L_lzma_crc64_table$non_lazy_ptr:
