diff --git a/src/liblzma/lz/lz_encoder.c b/src/liblzma/lz/lz_encoder.c
index 488095ca..fd1be9a7 100644
--- a/src/liblzma/lz/lz_encoder.c
+++ b/src/liblzma/lz/lz_encoder.c
@@ -47,7 +47,7 @@
 /// Calculate certain match finder properties and validate the calculated
 /// values. This is as its own function, because *num_items is needed to
 /// calculate memory requirements in common/memory.c.
-extern uint32_t
+extern bool
 lzma_lz_encoder_hash_properties(lzma_match_finder match_finder,
 		uint32_t history_size, uint32_t *restrict hash_mask,
 		uint32_t *restrict hash_size_sum, uint32_t *restrict num_items)
diff --git a/src/liblzma/lz/lz_encoder.h b/src/liblzma/lz/lz_encoder.h
index 1e47af2d..b13e4b83 100644
--- a/src/liblzma/lz/lz_encoder.h
+++ b/src/liblzma/lz/lz_encoder.h
@@ -132,7 +132,7 @@ struct lzma_lz_encoder_s {
 
 
 /// Calculates
-extern uint32_t lzma_lz_encoder_hash_properties(lzma_match_finder match_finder,
+extern bool lzma_lz_encoder_hash_properties(lzma_match_finder match_finder,
 		uint32_t history_size, uint32_t *restrict hash_mask,
 		uint32_t *restrict hash_size_sum,
 		uint32_t *restrict num_items);
