diff --git a/m4/visibility.m4 b/m4/visibility.m4
index 52c7ef10..4a741c5e 100644
--- a/m4/visibility.m4
+++ b/m4/visibility.m4
@@ -1,7 +1,8 @@
 dnl SPDX-License-Identifier: FSFULLR
 
-# visibility.m4 serial 8
-dnl Copyright (C) 2005, 2008, 2010-2023 Free Software Foundation, Inc.
+# visibility.m4
+# serial 9
+dnl Copyright (C) 2005, 2008, 2010-2024 Free Software Foundation, Inc.
 dnl This file is free software; the Free Software Foundation
 dnl gives unlimited permission to copy and/or distribute it,
 dnl with or without modifications, as long as this notice is preserved.
@@ -33,18 +34,18 @@ AC_DEFUN([gl_VISIBILITY],
     dnl user has put into $CC $CFLAGS $CPPFLAGS.
     AC_CACHE_CHECK([whether the -Werror option is usable],
       [gl_cv_cc_vis_werror],
-      [gl_save_CFLAGS="$CFLAGS"
+      [gl_saved_CFLAGS="$CFLAGS"
        CFLAGS="$CFLAGS -Werror"
        AC_COMPILE_IFELSE(
          [AC_LANG_PROGRAM([[]], [[]])],
          [gl_cv_cc_vis_werror=yes],
          [gl_cv_cc_vis_werror=no])
-       CFLAGS="$gl_save_CFLAGS"
+       CFLAGS="$gl_saved_CFLAGS"
       ])
     dnl Now check whether visibility declarations are supported.
     AC_CACHE_CHECK([for simple visibility declarations],
       [gl_cv_cc_visibility],
-      [gl_save_CFLAGS="$CFLAGS"
+      [gl_saved_CFLAGS="$CFLAGS"
        CFLAGS="$CFLAGS -fvisibility=hidden"
        dnl We use the option -Werror and a function dummyfunc, because on some
        dnl platforms (Cygwin 1.7) the use of -fvisibility triggers a warning
@@ -70,7 +71,7 @@ AC_DEFUN([gl_VISIBILITY],
             [[]])],
          [gl_cv_cc_visibility=yes],
          [gl_cv_cc_visibility=no])
-       CFLAGS="$gl_save_CFLAGS"
+       CFLAGS="$gl_saved_CFLAGS"
       ])
     if test $gl_cv_cc_visibility = yes; then
       CFLAG_VISIBILITY="-fvisibility=hidden"
