diff --git a/src/common/sysdefs.h b/src/common/sysdefs.h
index 97a1c045..61ac0b40 100644
--- a/src/common/sysdefs.h
+++ b/src/common/sysdefs.h
@@ -102,11 +102,11 @@
 #	elif SIZEOF_SIZE_T == 8
 #		define SIZE_MAX UINT64_MAX
 #	else
-#		error sizeof(size_t) is not 32-bit or 64-bit
+#		error size_t is not 32-bit or 64-bit
 #	endif
 #endif
 #if SIZE_MAX != UINT32_MAX && SIZE_MAX != UINT64_MAX
-#	error sizeof(size_t) is not 32-bit or 64-bit
+#	error size_t is not 32-bit or 64-bit
 #endif
 
 #include <stdlib.h>
