diff --git a/src/liblzma/common/string_conversion.c b/src/liblzma/common/string_conversion.c
index ae13ac46..c899783c 100644
--- a/src/liblzma/common/string_conversion.c
+++ b/src/liblzma/common/string_conversion.c
@@ -1002,6 +1002,12 @@ extern LZMA_API(const char *)
 lzma_str_to_filters(const char *str, int *error_pos, lzma_filter *filters,
 		uint32_t flags, const lzma_allocator *allocator)
 {
+	// If error_pos isn't NULL, *error_pos must always be set.
+	// liblzma <= 5.4.6 and <= 5.6.1 have a bug and don't do this
+	// when str == NULL or filters == NULL or flags are unsupported.
+	if (error_pos != NULL)
+		*error_pos = 0;
+
 	if (str == NULL || filters == NULL)
 		return "Unexpected NULL pointer argument(s) "
 				"to lzma_str_to_filters()";
