diff --git a/THANKS b/THANKS
index 96d5b590..f127e510 100644
--- a/THANKS
+++ b/THANKS
@@ -21,6 +21,7 @@ has been important. :-) In alphabetical order:
   - Milo Casagrande
   - Marek Černocký
   - Tomer Chachamu
+  - Gabi Davar
   - Chris Donawa
   - Andrew Dudman
   - Markus Duft
@@ -98,6 +99,7 @@ has been important. :-) In alphabetical order:
   - Loganaden Velvindron
   - Patrick J. Volkerding
   - Martin Väth
+  - Adam Walling
   - Christian Weisgerber
   - Bert Wesarg
   - Fredrik Wikstrom
