diff --git a/src/liblzma/api/lzma/base.h b/src/liblzma/api/lzma/base.h
index 53cf89f5..e7edf531 100644
--- a/src/liblzma/api/lzma/base.h
+++ b/src/liblzma/api/lzma/base.h
@@ -315,7 +315,7 @@ typedef struct lzma_internal_s lzma_internal;
  * Application must not touch the `internal' pointer.
  */
 typedef struct {
-	uint8_t *next_in;   /**< Pointer to the next input byte. */
+	const uint8_t *next_in; /**< Pointer to the next input byte. */
 	size_t avail_in;    /**< Number of available input bytes in next_in. */
 	uint64_t total_in;  /**< Total number of bytes read by liblzma. */
 
