diff --git a/THANKS b/THANKS
index 343233a6..d731a9b5 100644
--- a/THANKS
+++ b/THANKS
@@ -54,6 +54,7 @@ has been important. :-) In alphabetical order:
   - Paul Eggert
   - Robert Elz
   - Gilles Espinasse
+  - François Etcheverry
   - Denis Excoffier
   - Vincent Fazio
   - Michael Felt
@@ -105,6 +106,7 @@ has been important. :-) In alphabetical order:
   - Christian Kujau
   - Stephan Kulow
   - Ilya Kurdyukov
+  - Émilie Labbé
   - Peter Lawler
   - James M Leddy
   - Kelvin Lee
