diff --git a/THANKS b/THANKS
index 32af4568..ba5648f8 100644
--- a/THANKS
+++ b/THANKS
@@ -64,7 +64,10 @@ has been important. :-) In alphabetical order:
   - Juan Manuel Guerrero
   - Gabriela Gutierrez
   - Diederik de Haas
+  - Jan Terje Hansen
+  - Tobias Lahrmann Hansen
   - Joachim Henke
+  - Lizandro Heredia
   - Christian Hesse
   - Vincenzo Innocente
   - Peter Ivanov
@@ -111,6 +114,7 @@ has been important. :-) In alphabetical order:
   - Pierre-Yves Martin
   - Ed Maste
   - Martin Matuška
+  - Scott McAllister
   - Chris McCrohan
   - Ivan A. Melnikov
   - Jim Meyering
@@ -170,6 +174,7 @@ has been important. :-) In alphabetical order:
   - Dan Shechter
   - Stuart Shelton
   - Sebastian Andrzej Siewior
+  - Andrej Skenderija
   - Ville Skyttä
   - Brad Smith
   - Bruce Stark
