diff --git a/src/xz/args.c b/src/xz/args.c
index dff35f93..b0a1174a 100644
--- a/src/xz/args.c
+++ b/src/xz/args.c
@@ -213,7 +213,7 @@ parse_real(args_info *args, int argc, char **argv)
 		{ "ia64",         optional_argument, NULL,  OPT_IA64 },
 		{ "arm",          optional_argument, NULL,  OPT_ARM },
 		{ "armthumb",     optional_argument, NULL,  OPT_ARMTHUMB },
-		{ "experimental-arm64",        optional_argument, NULL,  OPT_ARM64 },
+		{ "arm64",        optional_argument, NULL,  OPT_ARM64 },
 		{ "sparc",        optional_argument, NULL,  OPT_SPARC },
 		{ "delta",        optional_argument, NULL,  OPT_DELTA },
 
