diff --git a/build-aux/ci_build.sh b/build-aux/ci_build.sh
index da780f33..d40cd3ce 100755
--- a/build-aux/ci_build.sh
+++ b/build-aux/ci_build.sh
@@ -48,6 +48,8 @@ DEST_DIR="$SRC_DIR/../xz_build"
 PHASE="all"
 ARTIFACTS_DIR_NAME="output"
 
+[[ -z ${CPU_COUNT} ]] && { CPU_COUNT=$(nproc 2>/dev/null || sysctl -n hw.ncpu); }
+[[ -z ${MAKEFLAGS} ]] && export MAKEFLAGS="-j${CPU_COUNT} -l${CPU_COUNT}"
 [[ -z ${CFLAGS} ]] && export CFLAGS="-O2"
 
 ###################
