diff --git a/src/liblzma/check/crc64_table.c b/src/liblzma/check/crc64_table.c
index 241adcd4..6f9adb04 100644
--- a/src/liblzma/check/crc64_table.c
+++ b/src/liblzma/check/crc64_table.c
@@ -20,8 +20,7 @@
 		|| (defined(__e2k__) && __iset__ >= 6)
 // No table needed but something has to be exported to keep some toolchains
 // happy. Also use a declaration to silence compiler warnings.
-extern const char lzma_crc64_dummy;
-const char lzma_crc64_dummy;
+typedef void lzma_crc64_dummy;
 
 #else
 // Having the declaration here silences clang -Wmissing-variable-declarations.
