diff --git a/src/lzma/help.c b/src/lzma/help.c
index f2a30ff6..85d754ec 100644
--- a/src/lzma/help.c
+++ b/src/lzma/help.c
@@ -97,6 +97,10 @@ show_help(void)
 "  --armthumb          ARM-Thumb filter\n"
 "  --sparc             SPARC filter\n"
 "\n"
+"  --delta=[OPTS]      Delta filter; valid OPTS (valid values; default):\n"
+"                        distance=NUM  Distance between bytes being\n"
+"                                      subtracted from each other (1-256; 1)\n"
+"\n"
 "  --copy              No filtering (useful only when specified alone)\n"
 "  --subblock=[OPTS]   Subblock filter; valid OPTS (valid values; default):\n"
 "                        size=NUM    number of bytes of data per subblock\n"
