diff --git a/src/xz/message.c b/src/xz/message.c
index ff218c48..19734e5b 100644
--- a/src/xz/message.c
+++ b/src/xz/message.c
@@ -896,7 +896,7 @@ uint32_to_optstr(uint32_t num)
 
 
 extern const char *
-message_filters_get(const lzma_filter *filters, bool all_known)
+message_filters_to_str(const lzma_filter *filters, bool all_known)
 {
 	static char buf[512];
 
@@ -1036,7 +1036,7 @@ message_filters_show(enum message_verbosity v, const lzma_filter *filters)
 		return;
 
 	fprintf(stderr, _("%s: Filter chain: %s\n"), progname,
-			message_filters_get(filters, true));
+			message_filters_to_str(filters, true));
 	return;
 }
 
diff --git a/src/xz/message.h b/src/xz/message.h
index 8f011874..aea4fdfd 100644
--- a/src/xz/message.h
+++ b/src/xz/message.h
@@ -94,7 +94,7 @@ extern void message_mem_needed(enum message_verbosity v, uint64_t memusage);
 ///                         into .xz headers are printed.
 ///
 /// \return     Pointer to a statically allocated buffer.
-extern const char *message_filters_get(
+extern const char *message_filters_to_str(
 		const lzma_filter *filters, bool all_known);
 
 
