diff --git a/tests/files/README b/tests/files/README
index f2b274c2..542a7b27 100644
--- a/tests/files/README
+++ b/tests/files/README
@@ -59,12 +59,27 @@
 
 2.2. Bad Files
 
-    bad-single-data_after_eopm.lzma has LZMA+Subblock, where the Subblock
+    bad-single-none-truncated.lzma is good-single-none.lzma without the
+    last byte of the file.
+
+    bad-cat-single-none-pad_garbage_1.lzma is good-cat-single-none-pad.lzma
+    with 0xFE appended to the end of the file. 0xFE doesn't begin .lzma
+    or LZMA_Alone format file.
+
+    bad-cat-single-none-pad_garbage_2.lzma is good-cat-single-none-pad.lzma
+    with 0xFF appended to the end of the file. 0xFF begins .lzma format
+    file, thus the decoder has to detect that the file is incomplete.
+
+    bad-cat-single-none-pad_garbage_3.lzma is good-cat-single-none-pad.lzma
+    with 0x5D appended to the end of the file. 0x5D is the most common
+    first byte of LZMA_Alone format file.
+
+    bad-single-data_after_eopm_1.lzma has LZMA+Subblock, where the Subblock
     filter gives one byte of data to LZMA after LZMA has detected EOPM.
 
     bad-single-data_after_eopm_2.lzma is like
-    bad-single-data_after_eopm.lzma but Subblock gives 256 MiB of data to
-    LZMA after LZMA has detected EOPM.
+    bad-single-data_after_eopm_1.lzma but Subblock gives 256 MiB of data
+    to LZMA after LZMA has detected EOPM.
 
     bad-single-subblock_subblock.lzma has Subblock+Subblock, where the
     Subblock decoder is given End of Input in the middle of a Subblock.
diff --git a/tests/files/bad-cat-single-none-pad_garbage_1.lzma b/tests/files/bad-cat-single-none-pad_garbage_1.lzma
new file mode 100644
index 00000000..447f1998
Binary files /dev/null and b/tests/files/bad-cat-single-none-pad_garbage_1.lzma differ
diff --git a/tests/files/bad-cat-single-none-pad_garbage_2.lzma b/tests/files/bad-cat-single-none-pad_garbage_2.lzma
new file mode 100644
index 00000000..26595aae
Binary files /dev/null and b/tests/files/bad-cat-single-none-pad_garbage_2.lzma differ
diff --git a/tests/files/bad-cat-single-none-pad_garbage_3.lzma b/tests/files/bad-cat-single-none-pad_garbage_3.lzma
new file mode 100644
index 00000000..73c87449
Binary files /dev/null and b/tests/files/bad-cat-single-none-pad_garbage_3.lzma differ
diff --git a/tests/files/bad-single-data_after_eopm.lzma b/tests/files/bad-single-data_after_eopm_1.lzma
similarity index 100%
rename from tests/files/bad-single-data_after_eopm.lzma
rename to tests/files/bad-single-data_after_eopm_1.lzma
diff --git a/tests/files/bad-single-none-truncated.lzma b/tests/files/bad-single-none-truncated.lzma
new file mode 100644
index 00000000..cc7aa56b
Binary files /dev/null and b/tests/files/bad-single-none-truncated.lzma differ
