diff --git a/build-aux/ci_build.sh b/build-aux/ci_build.sh
index 8c309028..6e264769 100755
--- a/build-aux/ci_build.sh
+++ b/build-aux/ci_build.sh
@@ -128,7 +128,7 @@ if [ "$PHASE" = "all" ] || [ "$PHASE" = "build" ]; then
 		fi
 
 		# Run configure script
-		"$SRC_DIR"/configure --enable-checks=$CHECK_TYPE $EXTRA_OPTIONS
+		"$SRC_DIR"/configure --enable-werror --enable-checks=$CHECK_TYPE $EXTRA_OPTIONS
 
 		# Build the project
 		make
