diff --git a/src/liblzma/validate_map.sh b/src/liblzma/validate_map.sh
index 5adf70b8..dd1589d2 100644
--- a/src/liblzma/validate_map.sh
+++ b/src/liblzma/validate_map.sh
@@ -122,7 +122,7 @@ DUPS=$(sort liblzma_generic.map | sed '/^$/d;/^global:$/d' | uniq -d)
 # ignored (@XZ_5.1.2alpha or @XZ_5.2.2 won't be added at all when
 # the #define HAVE_SYMBOL_VERSIONS_LINUX isn't used).
 IN_SYNC=
-if ! sed '109,123d' liblzma_linux.map \
+if ! sed '111,125d' liblzma_linux.map \
 		| cmp -s - liblzma_generic.map; then
 	IN_SYNC=no
 fi
