diff --git a/src/liblzma/common/memcmplen.h b/src/liblzma/common/memcmplen.h
index 4d1ef7b4..3c12422b 100644
--- a/src/liblzma/common/memcmplen.h
+++ b/src/liblzma/common/memcmplen.h
@@ -70,11 +70,11 @@ lzma_memcmplen(const uint8_t *buf1, const uint8_t *buf2,
 		const uint64_t x = read64ne(buf1 + len) - read64ne(buf2 + len);
 		if (x != 0) {
 	// MSVC or Intel C compiler on Windows
-#	if defined(_M_X64) && defined(_MSC_VER)
+#	if (defined(_MSC_VER) || defined(__INTEL_COMPILER)) && defined(_M_X64)
 			unsigned long tmp;
 			_BitScanForward64(&tmp, x);
 			len += (uint32_t)tmp >> 3;
-	// GCC, clang, or Intel C compiler
+	// GCC, Clang, or Intel C compiler
 #	else
 			len += (uint32_t)__builtin_ctzll(x) >> 3;
 #	endif
