diff --git a/src/liblzma/common/alone_decoder.c b/src/liblzma/common/alone_decoder.c
index 092047b4..91df5bf2 100644
--- a/src/liblzma/common/alone_decoder.c
+++ b/src/liblzma/common/alone_decoder.c
@@ -75,7 +75,8 @@ alone_decode(lzma_coder *coder,
 						& (UINT32_C(1) << i))
 					++count;
 
-			if (count > 1)
+			if (count != 1 || coder->options.lzma.dictionary_size
+					> LZMA_DICTIONARY_SIZE_MAX)
 				return LZMA_DATA_ERROR;
 
 			coder->pos = 0;
