diff --git a/src/xz/main.c b/src/xz/main.c
index 868d2888..4121f8d4 100644
--- a/src/xz/main.c
+++ b/src/xz/main.c
@@ -54,8 +54,8 @@ my_exit(enum exit_status_type status)
 			// If it was fclose() that failed, we have the reason
 			// in errno. If only ferror() indicated an error,
 			// we have no idea what the reason was.
-			message(V_ERROR, _("Writing to standard output "
-					"failed: %s"),
+			message(V_ERROR, "%s: %s", _("Writing to standard "
+						"output failed"),
 					fclose_err ? strerror(errno)
 						: _("Unknown error"));
 			status = E_ERROR;
diff --git a/src/xzdec/xzdec.c b/src/xzdec/xzdec.c
index 4827fc60..e99934f5 100644
--- a/src/xzdec/xzdec.c
+++ b/src/xzdec/xzdec.c
@@ -76,7 +76,7 @@ my_exit(void)
 		// If it was fclose() that failed, we have the reason
 		// in errno. If only ferror() indicated an error,
 		// we have no idea what the reason was.
-		my_errorf("Cannot write to standard output: %s", fclose_err
+		my_errorf("Writing to standard output failed: %s", fclose_err
 				? strerror(errno) : "Unknown error");
 		status = EXIT_FAILURE;
 	}
@@ -320,7 +320,7 @@ uncompress(lzma_stream *strm, FILE *file, const char *filename)
 	// FIXME: Maybe also LZMA_MEMLIMIT_ERROR in future?
 	if (ret != LZMA_OK) {
 		my_errorf("%s", ret == LZMA_MEM_ERROR ? strerror(ENOMEM)
-				: "Internal program error (bug)");
+				: "Internal error (bug)");
 		exit(EXIT_FAILURE);
 	}
 
@@ -427,7 +427,7 @@ uncompress(lzma_stream *strm, FILE *file, const char *filename)
 				break;
 
 			default:
-				msg = "Internal program error (bug)";
+				msg = "Internal error (bug)";
 				break;
 			}
 
