diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5dded28c..7f4a1b9e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -681,9 +681,11 @@ if(NOT MSVC AND HAVE_GETOPT_LONG)
             RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}"
                     COMPONENT xz)
 
-    install(FILES src/xz/xz.1
-            DESTINATION "${CMAKE_INSTALL_MANDIR}/man1"
-            COMPONENT xz)
+    if(UNIX)
+        install(FILES src/xz/xz.1
+                DESTINATION "${CMAKE_INSTALL_MANDIR}/man1"
+                COMPONENT xz)
+    endif()
 endif()
 
 
