diff --git a/src/xz/file_io.c b/src/xz/file_io.c
index d8a7597b..9c4d6ac8 100644
--- a/src/xz/file_io.c
+++ b/src/xz/file_io.c
@@ -895,7 +895,7 @@ io_open_dest_real(file_pair *pair)
 	// With fstat()/_fstat64() it works.
 	else if (pair->dest_fd != STDOUT_FILENO
 			&& !S_ISREG(pair->dest_st.st_mode)) {
-		message_error("%s: Destination is not a regular file",
+		message_error(_("%s: Destination is not a regular file"),
 				tuklib_mask_nonprint(pair->dest_name));
 
 		// dest_fd needs to be reset to -1 to keep io_close() working.
