diff --git a/windows/liblzma.vcxproj b/windows/liblzma.vcxproj
index cc85887e..2feafef9 100644
--- a/windows/liblzma.vcxproj
+++ b/windows/liblzma.vcxproj
@@ -285,13 +285,8 @@
     <ClInclude Include="..\src\common\tuklib_common.h" />
     <ClInclude Include="..\src\common\tuklib_config.h" />
     <ClInclude Include="..\src\common\tuklib_cpucores.h" />
-    <ClInclude Include="..\src\common\tuklib_exit.h" />
-    <ClInclude Include="..\src\common\tuklib_gettext.h" />
     <ClInclude Include="..\src\common\tuklib_integer.h" />
-    <ClInclude Include="..\src\common\tuklib_mbstr.h" />
-    <ClInclude Include="..\src\common\tuklib_open_stdxxx.h" />
     <ClInclude Include="..\src\common\tuklib_physmem.h" />
-    <ClInclude Include="..\src\common\tuklib_progname.h" />
     <ClInclude Include="..\src\liblzma\api\lzma.h" />
     <ClInclude Include="..\src\liblzma\api\lzma\base.h" />
     <ClInclude Include="..\src\liblzma\api\lzma\bcj.h" />
diff --git a/windows/liblzma_dll.vcxproj b/windows/liblzma_dll.vcxproj
index dd8a5853..92e98383 100644
--- a/windows/liblzma_dll.vcxproj
+++ b/windows/liblzma_dll.vcxproj
@@ -311,13 +311,8 @@
     <ClInclude Include="..\src\common\tuklib_common.h" />
     <ClInclude Include="..\src\common\tuklib_config.h" />
     <ClInclude Include="..\src\common\tuklib_cpucores.h" />
-    <ClInclude Include="..\src\common\tuklib_exit.h" />
-    <ClInclude Include="..\src\common\tuklib_gettext.h" />
     <ClInclude Include="..\src\common\tuklib_integer.h" />
-    <ClInclude Include="..\src\common\tuklib_mbstr.h" />
-    <ClInclude Include="..\src\common\tuklib_open_stdxxx.h" />
     <ClInclude Include="..\src\common\tuklib_physmem.h" />
-    <ClInclude Include="..\src\common\tuklib_progname.h" />
     <ClInclude Include="..\src\liblzma\api\lzma.h" />
     <ClInclude Include="..\src\liblzma\api\lzma\base.h" />
     <ClInclude Include="..\src\liblzma\api\lzma\bcj.h" />
