diff --git a/windows/build-with-cmake.bat b/windows/build-with-cmake.bat
index 24978ba8..f7a5b3f7 100644
--- a/windows/build-with-cmake.bat
+++ b/windows/build-with-cmake.bat
@@ -26,7 +26,7 @@ set PATH=%1;%2;%PATH%
 md build || exit /b
 cd build || exit /b
 
-cmake -G "MinGW Makefiles" -DCMAKE_BUILD_TYPE=Release -DENABLE_NLS=OFF -DBUILD_SHARED_LIBS=%3 ..\.. || exit /b
+cmake -G "MinGW Makefiles" -DCMAKE_BUILD_TYPE=Release -DXZ_NLS=OFF -DBUILD_SHARED_LIBS=%3 ..\.. || exit /b
 mingw32-make || exit /b
 mingw32-make test || exit /b
 
