diff --git a/README b/README
index 32d5084c..9d097def 100644
--- a/README
+++ b/README
@@ -298,7 +298,13 @@ XZ Utils
 6. Contact information
 ----------------------
 
-    Email: Lasse Collin <lasse.collin@tukaani.org>
-    IRC: Larhzu on #tukaani on Libera Chat
-    GitHub: https://github.com/tukaani-project/xz
+    XZ Utils in general:
+      - Home page: https://tukaani.org/xz/
+      - Email to maintainer(s): xz@tukaani.org
+      - IRC: #tukaani on Libera Chat
+      - GitHub: https://github.com/tukaani-project/xz
+
+    Lead maintainer:
+      - Email: Lasse Collin <lasse.collin@tukaani.org>
+      - IRC: Larhzu on Libera Chat
 
