diff --git a/TODO b/TODO
index af386101..f355e8c2 100644
--- a/TODO
+++ b/TODO
@@ -12,6 +12,10 @@ Known bugs
     it would be  possible by switching from BT2/BT3/BT4 match finder to
     HC3/HC4.
 
+    The code to detect number of CPU cores doesn't count hyperthreading
+    as multiple cores. In context of xz, it probably should.
+    Hyperthreading is good at least with p7zip.
+
 
 Missing features
 ----------------
