diff --git a/src/xz/args.c b/src/xz/args.c
index 835912ba..57f34823 100644
--- a/src/xz/args.c
+++ b/src/xz/args.c
@@ -81,11 +81,8 @@ parse_real(args_info *args, int argc, char **argv)
 		{ "lzma1",          optional_argument, NULL,  OPT_LZMA1 },
 		{ "lzma2",          optional_argument, NULL,  OPT_LZMA2 },
 		{ "x86",            optional_argument, NULL,  OPT_X86 },
-		{ "bcj",            optional_argument, NULL,  OPT_X86 },
 		{ "powerpc",        optional_argument, NULL,  OPT_POWERPC },
-		{ "ppc",            optional_argument, NULL,  OPT_POWERPC },
 		{ "ia64",           optional_argument, NULL,  OPT_IA64 },
-		{ "itanium",        optional_argument, NULL,  OPT_IA64 },
 		{ "arm",            optional_argument, NULL,  OPT_ARM },
 		{ "armthumb",       optional_argument, NULL,  OPT_ARMTHUMB },
 		{ "sparc",          optional_argument, NULL,  OPT_SPARC },
