diff --git a/src/lzmainfo/lzmainfo.c b/src/lzmainfo/lzmainfo.c
index 006a6b0f..caef7b50 100644
--- a/src/lzmainfo/lzmainfo.c
+++ b/src/lzmainfo/lzmainfo.c
@@ -14,13 +14,6 @@
 #include <stdio.h>
 #include <errno.h>
 
-#ifdef ENABLE_NLS
-#	include <libintl.h>
-#	define _(msgid) gettext(msgid)
-#else
-#	define _(msgid) msgid
-#endif
-
 #include "lzma.h"
 #include "getopt.h"
 #include "tuklib_gettext.h"
