diff --git a/src/liblzma/Makefile.am b/src/liblzma/Makefile.am
index d73d5f0a..97396c08 100644
--- a/src/liblzma/Makefile.am
+++ b/src/liblzma/Makefile.am
@@ -24,7 +24,7 @@ liblzma_la_CPPFLAGS = \
 	-I$(top_srcdir)/src/liblzma/simple \
 	-I$(top_srcdir)/src/common \
 	-DTUKLIB_SYMBOL_PREFIX=lzma_
-liblzma_la_LDFLAGS = -no-undefined -version-info 8:99:3
+liblzma_la_LDFLAGS = -no-undefined -version-info 9:0:4
 
 EXTRA_DIST += liblzma_generic.map liblzma_linux.map validate_map.sh
 if COND_SYMVERS_GENERIC
diff --git a/src/liblzma/api/lzma/version.h b/src/liblzma/api/lzma/version.h
index 1a80ba61..7d37130f 100644
--- a/src/liblzma/api/lzma/version.h
+++ b/src/liblzma/api/lzma/version.h
@@ -21,9 +21,9 @@
  * Version number split into components
  */
 #define LZMA_VERSION_MAJOR 5
-#define LZMA_VERSION_MINOR 3
-#define LZMA_VERSION_PATCH 5
-#define LZMA_VERSION_STABILITY LZMA_VERSION_STABILITY_BETA
+#define LZMA_VERSION_MINOR 4
+#define LZMA_VERSION_PATCH 0
+#define LZMA_VERSION_STABILITY LZMA_VERSION_STABILITY_STABLE
 
 #ifndef LZMA_VERSION_COMMIT
 #	define LZMA_VERSION_COMMIT ""
diff --git a/src/liblzma/liblzma_generic.map b/src/liblzma/liblzma_generic.map
index bba93c78..bb82167e 100644
--- a/src/liblzma/liblzma_generic.map
+++ b/src/liblzma/liblzma_generic.map
@@ -107,7 +107,7 @@ global:
 	lzma_stream_encoder_mt_memusage;
 } XZ_5.0;
 
-XZ_5.3.5beta {
+XZ_5.4 {
 global:
 	lzma_file_info_decoder;
 	lzma_filters_free;
diff --git a/src/liblzma/liblzma_linux.map b/src/liblzma/liblzma_linux.map
index 58c12ddb..449f5fd6 100644
--- a/src/liblzma/liblzma_linux.map
+++ b/src/liblzma/liblzma_linux.map
@@ -122,7 +122,7 @@ global:
 	lzma_stream_encoder_mt_memusage;
 } XZ_5.1.2alpha;
 
-XZ_5.3.5beta {
+XZ_5.4 {
 global:
 	lzma_file_info_decoder;
 	lzma_filters_free;
