diff --git a/README b/README
index c7d95013..7e878bfd 100644
--- a/README
+++ b/README
@@ -6,7 +6,7 @@ Important
 
     This is a beta version. The .xz file format is now stable though,
     which means that files created with the beta version will be
-    uncompressible with all future XZ Utils versions too (assuming
+    decompressible with all future XZ Utils versions too (assuming
     that there are no catastrophic bugs).
 
     liblzma API is pretty stable now, although minor tweaks may still
