diff --git a/README b/README
index eaeed4a8..ac812ff1 100644
--- a/README
+++ b/README
@@ -75,6 +75,7 @@ XZ Utils
     PACKAGERS             Information to packagers of XZ Utils
 
     COPYING               XZ Utils copyright and license information
+    COPYING.0BSD          BSD Zero Clause License
     COPYING.GPLv2         GNU General Public License version 2
     COPYING.GPLv3         GNU General Public License version 3
     COPYING.LGPLv2.1      GNU Lesser General Public License version 2.1
