diff --git a/src/xz/file_io.c b/src/xz/file_io.c
index f9807a69..61b10f98 100644
--- a/src/xz/file_io.c
+++ b/src/xz/file_io.c
@@ -511,14 +511,8 @@ io_open_src_real(file_pair *pair)
 #endif
 
 #ifdef HAVE_POSIX_FADVISE
-	const int fadvise_ret = posix_fadvise(
-			pair->src_fd, 0, 0, POSIX_FADV_SEQUENTIAL);
-
-	// It shouldn't fail, but if it does anyway, it doesn't matter.
-	// Check it with an assertion so that if something gets messed
-	// up in the future, it will get caught when debugging is enabled.
-	assert(fadvise_ret == 0);
-	(void)fadvise_ret;
+	// It will fail with some special files like FIFOs but that is fine.
+	(void)posix_fadvise(pair->src_fd, 0, 0, POSIX_FADV_SEQUENTIAL);
 #endif
 
 	return false;
