diff --git a/THANKS b/THANKS
index 7d2d4fe8..61e7a372 100644
--- a/THANKS
+++ b/THANKS
@@ -33,6 +33,7 @@ has been important. :-) In alphabetical order:
   - Tomer Chachamu
   - Vitaly Chikunov
   - Antoine Cœur
+  - Felix Collin
   - Gabi Davar
   - İhsan Doğan
   - Chris Donawa
@@ -123,6 +124,7 @@ has been important. :-) In alphabetical order:
   - Diego Elio Pettenò
   - Elbert Pol
   - Mikko Pouru
+  - Frank Prochnow
   - Rich Prohaska
   - Trần Ngọc Quân
   - Pavel Raiskup
@@ -157,6 +159,7 @@ has been important. :-) In alphabetical order:
   - Paul Townsend
   - Mohammed Adnène Trojette
   - Taiki Tsunekawa
+  - Mathieu Vachon
   - Maksym Vatsyk
   - Loganaden Velvindron
   - Patrick J. Volkerding
