diff --git a/CMakeLists.txt b/CMakeLists.txt
index f800396d..fab4fb71 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -161,6 +161,12 @@ set(CMAKE_C_STANDARD 11)
 # On Apple OSes, don't build executables as bundles:
 set(CMAKE_MACOSX_BUNDLE OFF)
 
+# The targets defined here don't support compiling as a unity build. Encoder
+# and decoder source files define different types with the same name, and some
+# internal header files don't have header guards leading to redeclaration
+# errors.
+set(CMAKE_UNITY_BUILD OFF)
+
 # Set CMAKE_INSTALL_LIBDIR and friends. This needs to be done before
 # the LOCALEDIR_DEFINITION workaround below.
 include(GNUInstallDirs)
