diff --git a/src/liblzma/check/crc32_x86.S b/src/liblzma/check/crc32_x86.S
index 859311ac..a9d07f04 100644
--- a/src/liblzma/check/crc32_x86.S
+++ b/src/liblzma/check/crc32_x86.S
@@ -262,17 +262,21 @@ LZMA_CRC32:
 #endif
 
 #if defined(__MACH__) && (defined(PIC) || defined(__PIC__))
+	/* Mach-O PIC */
 	.section __IMPORT,__pointers,non_lazy_symbol_pointers
 .L_lzma_crc32_table$non_lazy_ptr:
 	.indirect_symbol LZMA_CRC32_TABLE
 	.long 0
 
 #elif defined(_WIN32)
+#	ifndef LZMA_API_STATIC
 	/* This is equivalent of __declspec(dllexport). */
 	.section .drectve
 	.ascii " -export:lzma_crc32"
+#	endif
 
 #else
+	/* ELF */
 	.size	LZMA_CRC32, .-LZMA_CRC32
 #endif
 
diff --git a/src/liblzma/check/crc64_x86.S b/src/liblzma/check/crc64_x86.S
index abd8f9d4..031285e5 100644
--- a/src/liblzma/check/crc64_x86.S
+++ b/src/liblzma/check/crc64_x86.S
@@ -255,9 +255,11 @@ LZMA_CRC64:
 	.long 0
 
 #elif defined(_WIN32)
+#	ifndef LZMA_API_STATIC
 	/* This is equivalent of __declspec(dllexport). */
 	.section .drectve
 	.ascii " -export:lzma_crc64"
+#	endif
 
 #else
 	/* ELF */
