diff --git a/THANKS b/THANKS
index 787a13bf..7ad9dc7c 100644
--- a/THANKS
+++ b/THANKS
@@ -51,6 +51,7 @@ has been important. :-) In alphabetical order:
   - Gregory Margo
   - Jim Meyering
   - Arkadiusz Miskiewicz
+  - Conley Moorhous
   - Rafał Mużyło
   - Adrien Nader
   - Hongbo Ni
@@ -65,6 +66,7 @@ has been important. :-) In alphabetical order:
   - Pavel Raiskup
   - Robert Readman
   - Bernhard Reutner-Fischer
+  - Eric S. Raymond
   - Cristian Rodríguez
   - Christian von Roques
   - Jukka Salmi
