diff --git a/src/liblzma/api/lzma/version.h b/src/liblzma/api/lzma/version.h
index 5adddd2f..466c7822 100644
--- a/src/liblzma/api/lzma/version.h
+++ b/src/liblzma/api/lzma/version.h
@@ -22,7 +22,7 @@
  */
 #define LZMA_VERSION_MAJOR 5
 #define LZMA_VERSION_MINOR 3
-#define LZMA_VERSION_PATCH 2
+#define LZMA_VERSION_PATCH 3
 #define LZMA_VERSION_STABILITY LZMA_VERSION_STABILITY_ALPHA
 
 #ifndef LZMA_VERSION_COMMIT
diff --git a/src/liblzma/liblzma.map b/src/liblzma/liblzma.map
index 4462dac4..c37f293d 100644
--- a/src/liblzma/liblzma.map
+++ b/src/liblzma/liblzma.map
@@ -104,7 +104,7 @@ global:
 	lzma_stream_encoder_mt_memusage;
 } XZ_5.0;
 
-XZ_5.3.2alpha {
+XZ_5.3.3alpha {
 global:
 	lzma_microlzma_decoder;
 	lzma_microlzma_encoder;
