diff --git a/THANKS b/THANKS
index 571613db..aee1183a 100644
--- a/THANKS
+++ b/THANKS
@@ -156,6 +156,7 @@ has been important. :-) In alphabetical order:
   - Martin Storsjö
   - Jonathan Stott
   - Dan Stromberg
+  - Douglas Thor
   - Vincent Torri
   - Alexey Tourbin
   - Paul Townsend
@@ -181,6 +182,9 @@ has been important. :-) In alphabetical order:
   - Ryan Young
   - Andreas Zieringer
 
+Companies:
+  - Google
+
 Also thanks to all the people who have participated in the Tukaani project.
 
 I have probably forgot to add some names to the above list. Sorry about
