diff --git a/src/liblzma/common/stream_encoder_mt.c b/src/liblzma/common/stream_encoder_mt.c
index fb56a96f..3245aebd 100644
--- a/src/liblzma/common/stream_encoder_mt.c
+++ b/src/liblzma/common/stream_encoder_mt.c
@@ -1071,6 +1071,10 @@ stream_encoder_mt_init(lzma_next_coder *next, const lzma_allocator *allocator,
 	for (size_t i = 0; coder->filters[i].id != LZMA_VLI_UNKNOWN; ++i)
 		lzma_free(coder->filters[i].options, allocator);
 
+	// Mark it as empty so that it is in a safe state in case
+	// lzma_filters_copy() fails.
+	coder->filters[0].id = LZMA_VLI_UNKNOWN;
+
 	return_if_error(lzma_filters_copy(
 			filters, coder->filters, allocator));
 
