diff --git a/src/common/sysdefs.h b/src/common/sysdefs.h
index c74c6212..51f06cf0 100644
--- a/src/common/sysdefs.h
+++ b/src/common/sysdefs.h
@@ -65,6 +65,9 @@
 #ifndef PRIu32
 #	define PRIu32 "u"
 #endif
+#ifndef PRIx32
+#	define PRIx32 "x"
+#endif
 #ifndef PRIX32
 #	define PRIX32 "X"
 #endif
@@ -76,6 +79,9 @@
 #	ifndef PRIu64
 #		define PRIu64 "llu"
 #	endif
+#	ifndef PRIx64
+#		define PRIx64 "llx"
+#	endif
 #	ifndef PRIX64
 #		define PRIX64 "llX"
 #	endif
@@ -86,6 +92,9 @@
 #	ifndef PRIu64
 #		define PRIu64 "lu"
 #	endif
+#	ifndef PRIx64
+#		define PRIx64 "lx"
+#	endif
 #	ifndef PRIX64
 #		define PRIX64 "lX"
 #	endif
