diff --git a/configure.ac b/configure.ac
index c0400a93..4c34c719 100644
--- a/configure.ac
+++ b/configure.ac
@@ -26,7 +26,7 @@ AC_PREREQ(2.61)
 
 # [LZMA] instead of [LZMA utils] since I prefer to have lzma-version.tar.gz
 # instead of lzma-utils-version.tar.gz.
-AC_INIT([LZMA], [4.42.2alpha], [lasse.collin@tukaani.org])
+AC_INIT([LZMA], [4.42.3alpha], [lasse.collin@tukaani.org])
 
 AC_CONFIG_SRCDIR([src/liblzma/common/common.h])
 AC_CONFIG_HEADER([config.h])
diff --git a/src/liblzma/api/lzma/version.h b/src/liblzma/api/lzma/version.h
index ffbf8a81..5ccda502 100644
--- a/src/liblzma/api/lzma/version.h
+++ b/src/liblzma/api/lzma/version.h
@@ -31,7 +31,7 @@
  *    is incremented)
  *  - s indicates stability: 0 = alpha, 1 = beta, 2 = stable
  */
-#define LZMA_VERSION UINT32_C(40420010)
+#define LZMA_VERSION UINT32_C(40420030)
 
 
 /**
