diff --git a/windows/vs2017/liblzma.vcxproj b/windows/vs2017/liblzma.vcxproj
index a6b5f633..4941c2a2 100644
--- a/windows/vs2017/liblzma.vcxproj
+++ b/windows/vs2017/liblzma.vcxproj
@@ -230,6 +230,7 @@
     <ClCompile Include="..\..\src\liblzma\common\easy_encoder.c" />
     <ClCompile Include="..\..\src\liblzma\common\easy_encoder_memusage.c" />
     <ClCompile Include="..\..\src\liblzma\common\easy_preset.c" />
+    <ClCompile Include="..\..\src\liblzma\common\file_info.c" />
     <ClCompile Include="..\..\src\liblzma\common\filter_buffer_decoder.c" />
     <ClCompile Include="..\..\src\liblzma\common\filter_buffer_encoder.c" />
     <ClCompile Include="..\..\src\liblzma\common\filter_common.c" />
@@ -319,6 +320,7 @@
     <ClInclude Include="..\..\src\liblzma\common\filter_decoder.h" />
     <ClInclude Include="..\..\src\liblzma\common\filter_encoder.h" />
     <ClInclude Include="..\..\src\liblzma\common\index.h" />
+    <ClInclude Include="..\..\src\liblzma\common\index_decoder.h" />
     <ClInclude Include="..\..\src\liblzma\common\index_encoder.h" />
     <ClInclude Include="..\..\src\liblzma\common\memcmplen.h" />
     <ClInclude Include="..\..\src\liblzma\common\outqueue.h" />
diff --git a/windows/vs2017/liblzma_dll.vcxproj b/windows/vs2017/liblzma_dll.vcxproj
index dcea3037..5802f7d3 100644
--- a/windows/vs2017/liblzma_dll.vcxproj
+++ b/windows/vs2017/liblzma_dll.vcxproj
@@ -256,6 +256,7 @@
     <ClCompile Include="..\..\src\liblzma\common\easy_encoder.c" />
     <ClCompile Include="..\..\src\liblzma\common\easy_encoder_memusage.c" />
     <ClCompile Include="..\..\src\liblzma\common\easy_preset.c" />
+    <ClCompile Include="..\..\src\liblzma\common\file_info.c" />
     <ClCompile Include="..\..\src\liblzma\common\filter_buffer_decoder.c" />
     <ClCompile Include="..\..\src\liblzma\common\filter_buffer_encoder.c" />
     <ClCompile Include="..\..\src\liblzma\common\filter_common.c" />
@@ -345,6 +346,7 @@
     <ClInclude Include="..\..\src\liblzma\common\filter_decoder.h" />
     <ClInclude Include="..\..\src\liblzma\common\filter_encoder.h" />
     <ClInclude Include="..\..\src\liblzma\common\index.h" />
+    <ClInclude Include="..\..\src\liblzma\common\index_decoder.h" />
     <ClInclude Include="..\..\src\liblzma\common\index_encoder.h" />
     <ClInclude Include="..\..\src\liblzma\common\memcmplen.h" />
     <ClInclude Include="..\..\src\liblzma\common\outqueue.h" />
