diff --git a/NEWS b/NEWS
index 4f0c13fd..2d905d8b 100644
--- a/NEWS
+++ b/NEWS
@@ -164,6 +164,18 @@ XZ Utils Release Notes
       experimental and may change before it gets into a stable release.
 
 
+5.0.8 (2014-12-21)
+
+    * Fixed an old bug in xzgrep that affected OpenBSD and probably
+      a few other operating systems too.
+
+    * Updated French and German translations.
+
+    * Added support for detecting the amount of RAM on AmigaOS/AROS.
+
+    * Minor build system updates.
+
+
 5.0.7 (2014-09-20)
 
     * Fix regressions introduced in 5.0.6:
