diff --git a/src/liblzma/liblzma_w32res.rc b/src/liblzma/liblzma_w32res.rc
index e225b19e..19507b6e 100644
--- a/src/liblzma/liblzma_w32res.rc
+++ b/src/liblzma/liblzma_w32res.rc
@@ -5,7 +5,15 @@
  */
 
 #define MY_TYPE VFT_DLL
-#define MY_NAME "liblzma"
+
+#if defined(__MSYS__)
+#	define MY_NAME "msys-lzma-5"
+#elif defined(__CYGWIN__)
+#	define MY_NAME "cyglzma-5"
+#else
+#	define MY_NAME "liblzma"
+#endif
+
 #define MY_SUFFIX ".dll"
 #define MY_DESC "liblzma data compression library"
 #include "common_w32res.rc"
