diff --git a/CMakeLists.txt b/CMakeLists.txt
index 11928406..5756a6e0 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -175,14 +175,6 @@ endif()
 # to C99 if C11 isn't supported.
 set(CMAKE_C_STANDARD 11)
 
-# Support 32-bit x86 assembly files.
-if(NOT MSVC)
-    option(XZ_ASM_I386 "Enable 32-bit x86 assembly code" OFF)
-    if(XZ_ASM_I386)
-        enable_language(ASM)
-    endif()
-endif()
-
 # On Apple OSes, don't build executables as bundles:
 set(CMAKE_MACOSX_BUNDLE OFF)
 
@@ -249,6 +241,14 @@ add_compile_definitions(
     $<$<CONFIG:>:NDEBUG>
 )
 
+# Support 32-bit x86 assembly files.
+if(NOT MSVC)
+    option(XZ_ASM_I386 "Enable 32-bit x86 assembly code" OFF)
+    if(XZ_ASM_I386)
+        enable_language(ASM)
+    endif()
+endif()
+
 
 ######################
 # System definitions #
