diff --git a/CMakeLists.txt b/CMakeLists.txt
index 710dc455..8cf970fc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1850,7 +1850,7 @@ if(HAVE_DECODERS)
 
         if(WIN32)
             # Add the Windows resource file for xzdec.exe or lzmadec.exe.
-            target_sources("${XZDEC}" PRIVATE src/xzdec/xzdec_w32res.rc)
+            target_sources("${XZDEC}" PRIVATE "src/xzdec/${XZDEC}_w32res.rc")
             set_target_properties("${XZDEC}" PROPERTIES
                 LINK_DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/src/common/common_w32res.rc"
             )
