diff --git a/.gitignore b/.gitignore
index 206906c2..f6e57548 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,7 @@
+*~
+*.bak
+*.bak[0-9]
+
 .deps
 .libs
 *.la
@@ -10,20 +14,12 @@ Makefile.in
 /Doxyfile
 /aclocal.m4
 /build-aux
-/compile
-/config.guess
 /config.h
 /config.h.in
 /config.log
-/config.rpath
 /config.status
-/config.sub
 /configure
-/depcomp
-/install-sh
 /libtool
-/ltmain.sh
-/missing
 /stamp-h1
 
 /src/liblzma/liblzma.pc
@@ -31,22 +27,28 @@ Makefile.in
 /src/xzdec/lzmadec
 /src/xzdec/xzdec
 
+/src/scripts/xzdiff
+/src/scripts/xzgrep
+/src/scripts/xzless
+/src/scripts/xzmore
+
+/tests/compress_generated_abc
+/tests/compress_generated_random
+/tests/compress_generated_text
+/tests/create_compress_files
+/tests/test_block_header
+/tests/test_check
+/tests/test_filter_flags
+/tests/test_index
+/tests/test_stream_flags
+
 /lib/Makefile
 /tests/Makefile
 /Makefile
-/po/Makefile
 /debug/Makefile
 /src/scripts/Makefile
 /src/xz/Makefile
 /src/Makefile
-/src/liblzma/check/Makefile
-/src/liblzma/delta/Makefile
 /src/liblzma/Makefile
 /src/liblzma/api/Makefile
-/src/liblzma/lz/Makefile
-/src/liblzma/rangecoder/Makefile
-/src/liblzma/simple/Makefile
-/src/liblzma/lzma/Makefile
-/src/liblzma/common/Makefile
-/src/liblzma/subblock/Makefile
 /src/xzdec/Makefile
diff --git a/po/.gitignore b/po/.gitignore
index 3331837e..5db2e666 100644
--- a/po/.gitignore
+++ b/po/.gitignore
@@ -2,10 +2,13 @@ boldquot.sed
 en@boldquot.header
 en@quot.header
 insert-header.sin
+Makefile
+Makefile.in
 Makefile.in.in
 Makevars.template
 POTFILES
 quot.sed
+remove-potcdate.sed
 remove-potcdate.sin
 Rules-quot
 stamp-po
