diff --git a/README b/README
index 6b695d5c..cbe95578 100644
--- a/README
+++ b/README
@@ -207,8 +207,6 @@ XZ Utils
 
     If you have questions, bug reports, patches etc. related to XZ Utils,
     contact Lasse Collin <lasse.collin@tukaani.org> (in Finnish or English).
-    tukaani.org uses greylisting to reduce spam, thus when you send your
-    first email, it may get delayed by a few hours. In addition to that,
     I'm sometimes slow at replying. If you haven't got a reply within two
     weeks, assume that your email has got lost and resend it or use IRC.
 
