diff --git a/.gitignore b/.gitignore
index 3220269a..11dbc591 100644
--- a/.gitignore
+++ b/.gitignore
@@ -47,11 +47,15 @@ build-aux/test-driver
 /tests/compress_generated_random
 /tests/compress_generated_text
 /tests/create_compress_files
+/tests/test_bcj_exact_size
 /tests/test_block_header
 /tests/test_check
 /tests/test_filter_flags
 /tests/test_index
 /tests/test_stream_flags
+/tests/xzgrep_test_1.xz
+/tests/xzgrep_test_2.xz
+/tests/xzgrep_test_output
 
 /lib/Makefile
 /tests/Makefile
