diff --git a/AUTHORS b/AUTHORS
index d55d0518..69bbfc3e 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -19,6 +19,18 @@ Authors of XZ Utils
     Andrew Dudman helped adapting the scripts and their man pages for
     XZ Utils.
 
+    The initial version of the threaded .xz decompressor was written
+    by Sebastian Andrzej Siewior.
+
+    The initial version of the .lz (lzip) decoder was written
+    by Michał Górny.
+
+    CLMUL-accelerated CRC code was contributed by Ilya Kurdyukov.
+
+    Other authors:
+      - Jonathan Nieder
+      - Joachim Henke
+
     The GNU Autotools-based build system contains files from many authors,
     which I'm not trying to list here.
 
