diff --git a/src/liblzma/rangecoder/range_decoder.h b/src/liblzma/rangecoder/range_decoder.h
index eec721ac..a8ad5f02 100644
--- a/src/liblzma/rangecoder/range_decoder.h
+++ b/src/liblzma/rangecoder/range_decoder.h
@@ -37,7 +37,7 @@
 // The default can be overridden at build time by defining
 // LZMA_RANGE_DECODER_CONFIG to the desired mask.
 #ifndef LZMA_RANGE_DECODER_CONFIG
-#	if defined(__x86_64__) && (defined(__GNUC__) || defined(__clang__))
+#	if defined(__x86_64__) && !defined(__ILP32__) && (defined(__GNUC__) || defined(__clang__))
 #		define LZMA_RANGE_DECODER_CONFIG 0x1F0
 #	else
 #		define LZMA_RANGE_DECODER_CONFIG 0x03
