diff --git a/configure.ac b/configure.ac
index 5247c979..37949a94 100644
--- a/configure.ac
+++ b/configure.ac
@@ -716,13 +716,13 @@ AC_CONFIG_FILES([
 	src/xzdec/Makefile
 	src/lzmainfo/Makefile
 	src/scripts/Makefile
-	src/scripts/xzdiff
-	src/scripts/xzgrep
-	src/scripts/xzmore
-	src/scripts/xzless
 	tests/Makefile
 	debug/Makefile
 ])
+AC_CONFIG_FILES([src/scripts/xzdiff], [chmod +x src/scripts/xzdiff])
+AC_CONFIG_FILES([src/scripts/xzgrep], [chmod +x src/scripts/xzgrep])
+AC_CONFIG_FILES([src/scripts/xzmore], [chmod +x src/scripts/xzmore])
+AC_CONFIG_FILES([src/scripts/xzless], [chmod +x src/scripts/xzless])
 
 AC_OUTPUT
 
diff --git a/tests/test_scripts.sh b/tests/test_scripts.sh
index 891fc76a..293929e6 100755
--- a/tests/test_scripts.sh
+++ b/tests/test_scripts.sh
@@ -13,7 +13,7 @@
 XZ=../src/xz/xz
 XZDIFF=../src/scripts/xzdiff
 test -x "$XZ" || XZ=
-test -r "$XZDIFF" || XZDIFF=
+test -x "$XZDIFF" || XZDIFF=
 if test -z "$XZ" || test -z "$XZDIFF"; then
 	(exit 77)
 	exit 77
@@ -26,7 +26,7 @@ preimage=$srcdir/files/good-1-check-crc32.xz
 samepostimage=$srcdir/files/good-1-check-crc64.xz
 otherpostimage=$srcdir/files/good-1-lzma2-1.xz
 
-sh "$XZDIFF" "$preimage" "$samepostimage" >/dev/null
+"$XZDIFF" "$preimage" "$samepostimage" >/dev/null
 status=$?
 if test "$status" != 0 ; then
 	echo "xzdiff with no changes exited with status $status != 0"
@@ -34,7 +34,7 @@ if test "$status" != 0 ; then
 	exit 1
 fi
 
-sh "$XZDIFF" "$preimage" "$otherpostimage" >/dev/null
+"$XZDIFF" "$preimage" "$otherpostimage" >/dev/null
 status=$?
 if test "$status" != 1 ; then
 	echo "xzdiff with changes exited with status $status != 1"
@@ -42,7 +42,7 @@ if test "$status" != 1 ; then
 	exit 1
 fi
 
-sh "$XZDIFF" "$preimage" "$srcdir/files/missing.xz" >/dev/null 2>&1
+"$XZDIFF" "$preimage" "$srcdir/files/missing.xz" >/dev/null 2>&1
 status=$?
 if test "$status" != 2 ; then
 	echo "xzdiff with missing operand exited with status $status != 2"
