diff --git a/po/Makevars b/po/Makevars
index 9d82a84d..8f97f7de 100644
--- a/po/Makevars
+++ b/po/Makevars
@@ -14,8 +14,7 @@ subdir = po
 top_builddir = ..
 
 # These options get passed to xgettext.
-XGETTEXT_OPTIONS = --keyword=_ --keyword=N_ \
-	--add-location=file --no-wrap --package-name='XZ Utils'
+XGETTEXT_OPTIONS = --keyword=_ --keyword=N_ --add-location=file --no-wrap
 
 # This is the copyright holder that gets inserted into the header of the
 # $(DOMAIN).pot file.  Set this to the copyright holder of the surrounding
