diff --git a/THANKS b/THANKS
index e25f6935..787a13bf 100644
--- a/THANKS
+++ b/THANKS
@@ -6,6 +6,7 @@ Some people have helped more, some less, but nevertheless everyone's help
 has been important. :-) In alphabetical order:
   - Mark Adler
   - H. Peter Anvin
+  - Jeff Bastian
   - Nelson H. F. Beebe
   - Karl Berry
   - Anders F. Björklund
@@ -61,6 +62,7 @@ has been important. :-) In alphabetical order:
   - Diego Elio Pettenò
   - Elbert Pol
   - Mikko Pouru
+  - Pavel Raiskup
   - Robert Readman
   - Bernhard Reutner-Fischer
   - Cristian Rodríguez
