diff --git a/src/xz/args.c b/src/xz/args.c
index bb6e27bb..393ba6aa 100644
--- a/src/xz/args.c
+++ b/src/xz/args.c
@@ -222,6 +222,10 @@ parse_real(args_info *args, int argc, char **argv)
 		// --robot
 		case OPT_ROBOT:
 			opt_robot = true;
+
+			// This is to make sure that floating point numbers
+			// always have a dot as decimal separator.
+			setlocale(LC_NUMERIC, "C");
 			break;
 
 		case 'z':
