diff --git a/src/lzmainfo/Makefile.am b/src/lzmainfo/Makefile.am
index 3e18c5fb..b90b3b3d 100644
--- a/src/lzmainfo/Makefile.am
+++ b/src/lzmainfo/Makefile.am
@@ -12,6 +12,10 @@ lzmainfo_SOURCES = \
 	$(top_srcdir)/src/common/tuklib_progname.c \
 	$(top_srcdir)/src/common/tuklib_exit.c
 
+if COND_W32
+lzmainfo_SOURCES += lzmainfo_w32res.rc
+endif
+
 lzmainfo_CPPFLAGS = \
 	-DLOCALEDIR=\"$(localedir)\" \
 	-I$(top_srcdir)/src/common \
@@ -28,3 +32,8 @@ lzmainfo_LDADD += $(LTLIBINTL)
 
 
 dist_man_MANS = lzmainfo.1
+
+# Windows resource compiler support
+.rc.o:
+	$(RC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) \
+		$(lzmainfo_CPPFLAGS) $(CPPFLAGS) $(RCFLAGS) -i $< -o $@
diff --git a/src/lzmainfo/lzmainfo_w32res.rc b/src/lzmainfo/lzmainfo_w32res.rc
new file mode 100644
index 00000000..378fbe69
--- /dev/null
+++ b/src/lzmainfo/lzmainfo_w32res.rc
@@ -0,0 +1,12 @@
+/*
+ * Author: Lasse Collin
+ *
+ * This file has been put into the public domain.
+ * You can do whatever you want with this file.
+ */
+
+#define MY_TYPE VFT_APP
+#define MY_NAME "lzmainfo"
+#define MY_SUFFIX ".exe"
+#define MY_DESC "lzmainfo shows information about .lzma files"
+#include "common_w32res.rc"
