diff --git a/THANKS b/THANKS
index 61e7a372..571613db 100644
--- a/THANKS
+++ b/THANKS
@@ -70,6 +70,7 @@ has been important. :-) In alphabetical order:
   - Hans Jansen
   - Jouk Jansen
   - Jun I Jin
+  - Christoph Junghans
   - Kiyoshi Kanazawa
   - Joona Kannisto
   - Per Øyvind Karlsen
@@ -108,6 +109,7 @@ has been important. :-) In alphabetical order:
   - Nathan Moinvaziri
   - Étienne Mollier
   - Conley Moorhous
+  - Andrew Murray
   - Rafał Mużyło
   - Adrien Nader
   - Evan Nemerson
