diff --git a/src/liblzma/lz/lz_encoder.c b/src/liblzma/lz/lz_encoder.c
index bc38cae2..629f9df2 100644
--- a/src/liblzma/lz/lz_encoder.c
+++ b/src/liblzma/lz/lz_encoder.c
@@ -179,10 +179,8 @@ lzma_lz_encoder_reset(lzma_lz_encoder *lz, lzma_allocator *allocator,
 		}
 	}
 
-	// Allocation successful. Store the new size and calculate
-	// must_move_pos.
+	// Allocation successful. Store the new size.
 	lz->size = buffer_size;
-	lz->must_move_pos = lz->size - lz->keep_size_after;
 
 	// Reset in window variables.
 	lz->offset = 0;
@@ -373,7 +371,7 @@ fill_window(lzma_coder *coder, lzma_allocator *allocator, const uint8_t *in,
 	lzma_ret ret;
 
 	// Move the sliding window if needed.
-	if (coder->lz.read_pos >= coder->lz.must_move_pos)
+	if (coder->lz.read_pos >= coder->lz.size - coder->lz.keep_size_after)
 		move_window(&coder->lz);
 
 	if (coder->next.code == NULL) {
diff --git a/src/liblzma/lz/lz_encoder.h b/src/liblzma/lz/lz_encoder.h
index b39c88e5..fe94618b 100644
--- a/src/liblzma/lz/lz_encoder.h
+++ b/src/liblzma/lz/lz_encoder.h
@@ -72,10 +72,6 @@ struct lzma_lz_encoder_s {
 	/// to buffer[write_pos].
 	size_t write_pos;
 
-	/// When read_pos >= must_move_pos, move_window() must be called
-	/// to make more space for the input data.
-	size_t must_move_pos;
-
 	/// Number of bytes that must be kept available in our input history.
 	/// That is, once keep_size_before bytes have been processed,
 	/// buffer[read_pos - keep_size_before] is the oldest byte that
