diff --git a/THANKS b/THANKS
index 5f607a2f..bee56b11 100644
--- a/THANKS
+++ b/THANKS
@@ -12,6 +12,7 @@ has been important. :-) In alphabetical order:
   - Emmanuel Blot
   - Martin Blumenstingl
   - Jakub Bogusz
+  - Maarten Bosmans
   - Trent W. Buck
   - David Burklund
   - Daniel Mealha Cabrita
@@ -38,6 +39,7 @@ has been important. :-) In alphabetical order:
   - Peter Lawler
   - Hin-Tak Leung
   - Andraž 'ruskie' Levstik
+  - Wim Lewis
   - Lorenzo De Liso
   - Gregory Margo
   - Jim Meyering
@@ -61,6 +63,7 @@ has been important. :-) In alphabetical order:
   - Dan Shechter
   - Stuart Shelton
   - Jonathan Stott
+  - Dan Stromberg
   - Paul Townsend
   - Mohammed Adnène Trojette
   - Alexey Tourbin
