diff --git a/tests/files/README b/tests/files/README
index 734b828b..59b54e42 100644
--- a/tests/files/README
+++ b/tests/files/README
@@ -93,6 +93,11 @@
 
     good-1-3delta-lzma2.xz has three Delta filters and LZMA2.
 
+    good-1-empty-bcj-lzma2.xz has an empty Block that uses PowerPC BCJ
+    and LZMA2. liblzma from XZ Utils 5.0.1 and older may incorrectly
+    return LZMA_BUF_ERROR in some cases. See commit message
+    d8db706acb8316f9861abd432cfbe001dd6d0c5c for the details.
+
 
 2.2. Unsupported Files
 
diff --git a/tests/files/good-1-empty-bcj-lzma2.xz b/tests/files/good-1-empty-bcj-lzma2.xz
new file mode 100644
index 00000000..94016d8b
Binary files /dev/null and b/tests/files/good-1-empty-bcj-lzma2.xz differ
