diff --git a/po/Makevars b/po/Makevars
index 627d34ec..3f56947c 100644
--- a/po/Makevars
+++ b/po/Makevars
@@ -63,7 +63,7 @@ USE_MSGCTXT = no
 # Useful options are in particular:
 #   --previous            to keep previous msgids of translated messages,
 #   --quiet               to reduce the verbosity.
-MSGMERGE_OPTIONS = --no-wrap
+MSGMERGE_OPTIONS = --add-location=file --no-wrap
 
 # These options get passed to msginit.
 # If you want to disable line wrapping when writing PO files, add
