diff --git a/src/xz/message.c b/src/xz/message.c
index c62e2b2c..4ccc893c 100644
--- a/src/xz/message.c
+++ b/src/xz/message.c
@@ -1076,8 +1076,11 @@ message_help(bool long_help)
 			"Compress or decompress FILEs in the .xz format.\n\n"),
 			progname);
 
-	puts(_("Mandatory arguments to long options are mandatory for "
-			"short options too.\n"));
+	// NOTE: The short help doesn't currently have options that
+	// take arguments.
+	if (long_help)
+		puts(_("Mandatory arguments to long options are mandatory "
+				"for short options too.\n"));
 
 	if (long_help)
 		puts(_(" Operation mode:\n"));
