diff --git a/THANKS b/THANKS
index cec8bbdd..426a2ed0 100644
--- a/THANKS
+++ b/THANKS
@@ -16,6 +16,7 @@ has been important. :-) In alphabetical order:
   - Per Øyvind Karlsen
   - Ville Koskinen
   - Stephan Kulow
+  - Peter Lawler
   - Hin-Tak Leung
   - Andraž 'ruskie' Levstik
   - Jim Meyering
@@ -24,6 +25,7 @@ has been important. :-) In alphabetical order:
   - Igor Pavlov
   - Mikko Pouru
   - Bernhard Reutner-Fischer
+  - Christian von Roques
   - Alexandre Sauvé
   - Andreas Schwab
   - Dan Shechter
