diff --git a/tests/files/README b/tests/files/README
index 53950edb..1ee81bda 100644
--- a/tests/files/README
+++ b/tests/files/README
@@ -209,6 +209,13 @@
 
     bad-1-check-crc32.xz has wrong Check (CRC32).
 
+    bad-1-check-crc32-2.xz has Compressed Size and Uncompressed Size in
+    Block Header but wrong Check (CRC32) in the actual data. This file
+    differs by one byte from good-1-block_header-1.xz: the last byte of
+    the Check field is wrong. This file is useful for testing error
+    detection in the threaded decoder when a worker thread is configured
+    to pass input one byte at a time to the Block decoder.
+
     bad-1-check-crc64.xz has wrong Check (CRC64).
 
     bad-1-check-sha256.xz has wrong Check (SHA-256).
diff --git a/tests/files/bad-1-check-crc32-2.xz b/tests/files/bad-1-check-crc32-2.xz
new file mode 100644
index 00000000..95731ae3
Binary files /dev/null and b/tests/files/bad-1-check-crc32-2.xz differ
