diff --git a/windows/Makefile b/windows/Makefile
index c6ebfc94..2da34506 100644
--- a/windows/Makefile
+++ b/windows/Makefile
@@ -132,6 +132,8 @@ LIBLZMA_SRCS_C = \
 	../src/liblzma/common/easy_encoder.c \
 	../src/liblzma/common/easy_encoder_memusage.c \
 	../src/liblzma/common/easy_preset.c \
+	../src/liblzma/common/filter_buffer_decoder.c \
+	../src/liblzma/common/filter_buffer_encoder.c \
 	../src/liblzma/common/filter_common.c \
 	../src/liblzma/common/filter_decoder.c \
 	../src/liblzma/common/filter_encoder.c \
